<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>6 Support vector machines | An Introduction to Machine Learning</title>
  <meta name="description" content="Course materials for An Introduction to Machine Learning" />
  <meta name="generator" content="bookdown 0.27 and GitBook 2.6.7" />

  <meta property="og:title" content="6 Support vector machines | An Introduction to Machine Learning" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="/figures/cover_image.png" />
  <meta property="og:description" content="Course materials for An Introduction to Machine Learning" />
  <meta name="github-repo" content="bioinformatics-training/intro-machine-learning-2019" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="6 Support vector machines | An Introduction to Machine Learning" />
  
  <meta name="twitter:description" content="Course materials for An Introduction to Machine Learning" />
  <meta name="twitter:image" content="/figures/cover_image.png" />

<meta name="author" content="Sudhakaran Prabakaran, Matt Wayland and Chris Penfold" />


<meta name="date" content="2020-10-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="nearest-neighbours.html"/>
<link rel="next" href="decision-trees.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About the course</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#overview"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#registration"><i class="fa fa-check"></i><b>1.2</b> Registration</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.3</b> Prerequisites</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.4</b> Github</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.6</b> Contact</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#colophon"><i class="fa fa-check"></i><b>1.7</b> Colophon</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#what-is-machine-learning"><i class="fa fa-check"></i><b>2.1</b> What is machine learning?</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#aspects-of-ml"><i class="fa fa-check"></i><b>2.2</b> Aspects of ML</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#what-actually-happened-under-the-hood"><i class="fa fa-check"></i><b>2.3</b> What actually happened under the hood</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#introduction-to-caret"><i class="fa fa-check"></i><b>2.4</b> Introduction to CARET</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="intro.html"><a href="intro.html#preprocessing-with-the-iris-dataset"><i class="fa fa-check"></i><b>2.4.1</b> Preprocessing with the Iris dataset</a></li>
<li class="chapter" data-level="2.4.2" data-path="intro.html"><a href="intro.html#training-different-types-of-models"><i class="fa fa-check"></i><b>2.4.2</b> Training different types of models</a></li>
<li class="chapter" data-level="2.4.3" data-path="intro.html"><a href="intro.html#cross-validation"><i class="fa fa-check"></i><b>2.4.3</b> Cross-validation</a></li>
<li class="chapter" data-level="2.4.4" data-path="intro.html"><a href="intro.html#optimising-hyperparameters"><i class="fa fa-check"></i><b>2.4.4</b> Optimising hyperparameters</a></li>
<li class="chapter" data-level="2.4.5" data-path="intro.html"><a href="intro.html#using-dummy-variables-with-the-sacramento-dataset"><i class="fa fa-check"></i><b>2.4.5</b> Using dummy variables with the Sacramento dataset</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>3</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="3.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#linear-dimensionality-reduction"><i class="fa fa-check"></i><b>3.1</b> Linear Dimensionality Reduction</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#interpreting-the-principle-component-axes"><i class="fa fa-check"></i><b>3.1.1</b> Interpreting the Principle Component Axes</a></li>
<li class="chapter" data-level="3.1.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#horseshoe-effect"><i class="fa fa-check"></i><b>3.1.2</b> Horseshoe effect</a></li>
<li class="chapter" data-level="3.1.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#pca-analysis-of-mammalian-development"><i class="fa fa-check"></i><b>3.1.3</b> PCA analysis of mammalian development</a></li>
<li class="chapter" data-level="3.1.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#biological-interpretation"><i class="fa fa-check"></i><b>3.1.4</b> Biological interpretation</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#nonlinear-dimensionality-reduction"><i class="fa fa-check"></i><b>3.2</b> Nonlinear Dimensionality Reduction</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#nonlinear-warping"><i class="fa fa-check"></i><b>3.2.1</b> Nonlinear warping</a></li>
<li class="chapter" data-level="3.2.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#stochasticity"><i class="fa fa-check"></i><b>3.2.2</b> Stochasticity</a></li>
<li class="chapter" data-level="3.2.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#analysis-of-mammalian-development"><i class="fa fa-check"></i><b>3.2.3</b> Analysis of mammalian development</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#other-dimensionality-reduction-techniques"><i class="fa fa-check"></i><b>3.3</b> Other dimensionality reduction techniques</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>4</b> Clustering</a>
<ul>
<li class="chapter" data-level="4.1" data-path="clustering.html"><a href="clustering.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="clustering.html"><a href="clustering.html#registered-s3-method-overwritten-by-ggally"><i class="fa fa-check"></i><b>4.2</b> Registered S3 method overwritten by ‘GGally’:</a></li>
<li class="chapter" data-level="4.3" data-path="clustering.html"><a href="clustering.html#method-from"><i class="fa fa-check"></i><b>4.3</b> method from</a></li>
<li class="chapter" data-level="4.4" data-path="clustering.html"><a href="clustering.html#gg-ggplot2"><i class="fa fa-check"></i><b>4.4</b> +.gg ggplot2</a></li>
<li class="chapter" data-level="4.5" data-path="clustering.html"><a href="clustering.html#distance-metrics"><i class="fa fa-check"></i><b>4.5</b> Distance metrics</a></li>
<li class="chapter" data-level="4.6" data-path="clustering.html"><a href="clustering.html#hierarchic-agglomerative"><i class="fa fa-check"></i><b>4.6</b> Hierarchic agglomerative</a>
<ul>
<li class="chapter" data-level="4.6.1" data-path="clustering.html"><a href="clustering.html#linkage-algorithms"><i class="fa fa-check"></i><b>4.6.1</b> Linkage algorithms</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="clustering.html"><a href="clustering.html#k-means"><i class="fa fa-check"></i><b>4.7</b> K-means</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="clustering.html"><a href="clustering.html#algorithm"><i class="fa fa-check"></i><b>4.7.1</b> Algorithm</a></li>
<li class="chapter" data-level="4.7.2" data-path="clustering.html"><a href="clustering.html#choosing-initial-cluster-centres"><i class="fa fa-check"></i><b>4.7.2</b> Choosing initial cluster centres</a></li>
<li class="chapter" data-level="4.7.3" data-path="clustering.html"><a href="clustering.html#choosingK"><i class="fa fa-check"></i><b>4.7.3</b> Choosing k</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="clustering.html"><a href="clustering.html#dbscan"><i class="fa fa-check"></i><b>4.8</b> DBSCAN</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="clustering.html"><a href="clustering.html#algorithm-1"><i class="fa fa-check"></i><b>4.8.1</b> Algorithm</a></li>
<li class="chapter" data-level="4.8.2" data-path="clustering.html"><a href="clustering.html#implementation-in-r"><i class="fa fa-check"></i><b>4.8.2</b> Implementation in R</a></li>
<li class="chapter" data-level="4.8.3" data-path="clustering.html"><a href="clustering.html#choosing-parameters"><i class="fa fa-check"></i><b>4.8.3</b> Choosing parameters</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="clustering.html"><a href="clustering.html#section"><i class="fa fa-check"></i><b>4.9</b> </a></li>
<li class="chapter" data-level="4.10" data-path="clustering.html"><a href="clustering.html#section-1"><i class="fa fa-check"></i><b>4.10</b> 0 1 2 3</a></li>
<li class="chapter" data-level="4.11" data-path="clustering.html"><a href="clustering.html#section-2"><i class="fa fa-check"></i><b>4.11</b> 43 484 486 487</a></li>
<li class="chapter" data-level="4.12" data-path="clustering.html"><a href="clustering.html#example-clustering-synthetic-data-sets"><i class="fa fa-check"></i><b>4.12</b> Example: clustering synthetic data sets</a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="clustering.html"><a href="clustering.html#hierarchic-agglomerative-1"><i class="fa fa-check"></i><b>4.12.1</b> Hierarchic agglomerative</a></li>
</ul></li>
<li class="chapter" data-level="4.13" data-path="clustering.html"><a href="clustering.html#section-3"><i class="fa fa-check"></i><b>4.13</b> </a></li>
<li class="chapter" data-level="4.14" data-path="clustering.html"><a href="clustering.html#section-4"><i class="fa fa-check"></i><b>4.14</b> ———————</a></li>
<li class="chapter" data-level="4.15" data-path="clustering.html"><a href="clustering.html#welcome-to-dendextend-version-1.15.2"><i class="fa fa-check"></i><b>4.15</b> Welcome to dendextend version 1.15.2</a></li>
<li class="chapter" data-level="4.16" data-path="clustering.html"><a href="clustering.html#type-citationdendextend-for-how-to-cite-the-package."><i class="fa fa-check"></i><b>4.16</b> Type citation(‘dendextend’) for how to cite the package.</a></li>
<li class="chapter" data-level="4.17" data-path="clustering.html"><a href="clustering.html#section-5"><i class="fa fa-check"></i><b>4.17</b> </a></li>
<li class="chapter" data-level="4.18" data-path="clustering.html"><a href="clustering.html#type-browsevignettespackage-dendextend-for-the-package-vignette."><i class="fa fa-check"></i><b>4.18</b> Type browseVignettes(package = ‘dendextend’) for the package vignette.</a></li>
<li class="chapter" data-level="4.19" data-path="clustering.html"><a href="clustering.html#the-github-page-is-httpsgithub.comtalgalilidendextend"><i class="fa fa-check"></i><b>4.19</b> The github page is: <span>https://github.com/talgalili/dendextend/</span></a></li>
<li class="chapter" data-level="4.20" data-path="clustering.html"><a href="clustering.html#section-6"><i class="fa fa-check"></i><b>4.20</b> </a></li>
<li class="chapter" data-level="4.21" data-path="clustering.html"><a href="clustering.html#suggestions-and-bug-reports-can-be-submitted-at-httpsgithub.comtalgalilidendextendissues"><i class="fa fa-check"></i><b>4.21</b> Suggestions and bug-reports can be submitted at: <span>https://github.com/talgalili/dendextend/issues</span></a></li>
<li class="chapter" data-level="4.22" data-path="clustering.html"><a href="clustering.html#you-may-ask-questions-at-stackoverflow-use-the-r-and-dendextend-tags"><i class="fa fa-check"></i><b>4.22</b> You may ask questions at stackoverflow, use the r and dendextend tags:</a></li>
<li class="chapter" data-level="4.23" data-path="clustering.html"><a href="clustering.html#httpsstackoverflow.comquestionstaggeddendextend"><i class="fa fa-check"></i><b>4.23</b> <span>https://stackoverflow.com/questions/tagged/dendextend</span></a></li>
<li class="chapter" data-level="4.24" data-path="clustering.html"><a href="clustering.html#section-7"><i class="fa fa-check"></i><b>4.24</b> </a></li>
<li class="chapter" data-level="4.25" data-path="clustering.html"><a href="clustering.html#to-suppress-this-message-use-suppresspackagestartupmessageslibrarydendextend"><i class="fa fa-check"></i><b>4.25</b> To suppress this message use: suppressPackageStartupMessages(library(dendextend))</a></li>
<li class="chapter" data-level="4.26" data-path="clustering.html"><a href="clustering.html#section-8"><i class="fa fa-check"></i><b>4.26</b> ———————</a></li>
<li class="chapter" data-level="4.27" data-path="clustering.html"><a href="clustering.html#section-9"><i class="fa fa-check"></i><b>4.27</b> </a></li>
<li class="chapter" data-level="4.28" data-path="clustering.html"><a href="clustering.html#attaching-package-dendextend"><i class="fa fa-check"></i><b>4.28</b> Attaching package: ‘dendextend’</a></li>
<li class="chapter" data-level="4.29" data-path="clustering.html"><a href="clustering.html#the-following-object-is-masked-from-packageggdendro"><i class="fa fa-check"></i><b>4.29</b> The following object is masked from ‘package:ggdendro’:</a></li>
<li class="chapter" data-level="4.30" data-path="clustering.html"><a href="clustering.html#section-10"><i class="fa fa-check"></i><b>4.30</b> </a></li>
<li class="chapter" data-level="4.31" data-path="clustering.html"><a href="clustering.html#theme_dendro"><i class="fa fa-check"></i><b>4.31</b> theme_dendro</a></li>
<li class="chapter" data-level="4.32" data-path="clustering.html"><a href="clustering.html#the-following-object-is-masked-from-packagestats"><i class="fa fa-check"></i><b>4.32</b> The following object is masked from ‘package:stats’:</a></li>
<li class="chapter" data-level="4.33" data-path="clustering.html"><a href="clustering.html#section-11"><i class="fa fa-check"></i><b>4.33</b> </a></li>
<li class="chapter" data-level="4.34" data-path="clustering.html"><a href="clustering.html#cutree"><i class="fa fa-check"></i><b>4.34</b> cutree</a>
<ul>
<li class="chapter" data-level="4.34.1" data-path="clustering.html"><a href="clustering.html#k-means-1"><i class="fa fa-check"></i><b>4.34.1</b> K-means</a></li>
</ul></li>
<li class="chapter" data-level="4.35" data-path="clustering.html"><a href="clustering.html#warning-did-not-converge-in-10-iterations"><i class="fa fa-check"></i><b>4.35</b> Warning: did not converge in 10 iterations</a></li>
<li class="chapter" data-level="4.36" data-path="clustering.html"><a href="clustering.html#warning-did-not-converge-in-10-iterations-1"><i class="fa fa-check"></i><b>4.36</b> Warning: did not converge in 10 iterations</a>
<ul>
<li class="chapter" data-level="4.36.1" data-path="clustering.html"><a href="clustering.html#dbscan-1"><i class="fa fa-check"></i><b>4.36.1</b> DBSCAN</a></li>
</ul></li>
<li class="chapter" data-level="4.37" data-path="clustering.html"><a href="clustering.html#section-12"><i class="fa fa-check"></i><b>4.37</b> </a></li>
<li class="chapter" data-level="4.38" data-path="clustering.html"><a href="clustering.html#section-13"><i class="fa fa-check"></i><b>4.38</b> 0 1 2 3 4 5 6</a></li>
<li class="chapter" data-level="4.39" data-path="clustering.html"><a href="clustering.html#section-14"><i class="fa fa-check"></i><b>4.39</b> 2 168 307 105 127 45 34</a></li>
<li class="chapter" data-level="4.40" data-path="clustering.html"><a href="clustering.html#section-15"><i class="fa fa-check"></i><b>4.40</b> </a></li>
<li class="chapter" data-level="4.41" data-path="clustering.html"><a href="clustering.html#section-16"><i class="fa fa-check"></i><b>4.41</b> 0 1 2</a></li>
<li class="chapter" data-level="4.42" data-path="clustering.html"><a href="clustering.html#section-17"><i class="fa fa-check"></i><b>4.42</b> 8 748 744</a></li>
<li class="chapter" data-level="4.43" data-path="clustering.html"><a href="clustering.html#section-18"><i class="fa fa-check"></i><b>4.43</b> </a></li>
<li class="chapter" data-level="4.44" data-path="clustering.html"><a href="clustering.html#section-19"><i class="fa fa-check"></i><b>4.44</b> 0 1</a></li>
<li class="chapter" data-level="4.45" data-path="clustering.html"><a href="clustering.html#section-20"><i class="fa fa-check"></i><b>4.45</b> 40 1460</a></li>
<li class="chapter" data-level="4.46" data-path="clustering.html"><a href="clustering.html#section-21"><i class="fa fa-check"></i><b>4.46</b> </a></li>
<li class="chapter" data-level="4.47" data-path="clustering.html"><a href="clustering.html#section-22"><i class="fa fa-check"></i><b>4.47</b> 0 1 2</a></li>
<li class="chapter" data-level="4.48" data-path="clustering.html"><a href="clustering.html#section-23"><i class="fa fa-check"></i><b>4.48</b> 109 399 992</a></li>
<li class="chapter" data-level="4.49" data-path="clustering.html"><a href="clustering.html#section-24"><i class="fa fa-check"></i><b>4.49</b> </a></li>
<li class="chapter" data-level="4.50" data-path="clustering.html"><a href="clustering.html#section-25"><i class="fa fa-check"></i><b>4.50</b> 0 1 2 3</a></li>
<li class="chapter" data-level="4.51" data-path="clustering.html"><a href="clustering.html#section-26"><i class="fa fa-check"></i><b>4.51</b> 29 489 488 494</a></li>
<li class="chapter" data-level="4.52" data-path="clustering.html"><a href="clustering.html#section-27"><i class="fa fa-check"></i><b>4.52</b> </a></li>
<li class="chapter" data-level="4.53" data-path="clustering.html"><a href="clustering.html#section-28"><i class="fa fa-check"></i><b>4.53</b> 1</a></li>
<li class="chapter" data-level="4.54" data-path="clustering.html"><a href="clustering.html#section-29"><i class="fa fa-check"></i><b>4.54</b> 1500</a></li>
<li class="chapter" data-level="4.55" data-path="clustering.html"><a href="clustering.html#evaluating-cluster-quality"><i class="fa fa-check"></i><b>4.55</b> Evaluating cluster quality</a>
<ul>
<li class="chapter" data-level="4.55.1" data-path="clustering.html"><a href="clustering.html#silhouetteMethod"><i class="fa fa-check"></i><b>4.55.1</b> Silhouette method</a></li>
<li class="chapter" data-level="4.55.2" data-path="clustering.html"><a href="clustering.html#example---k-means-clustering-of-blobs-data-set"><i class="fa fa-check"></i><b>4.55.2</b> Example - k-means clustering of blobs data set</a></li>
<li class="chapter" data-level="4.55.3" data-path="clustering.html"><a href="clustering.html#example---dbscan-clustering-of-noisy-moons"><i class="fa fa-check"></i><b>4.55.3</b> Example - DBSCAN clustering of noisy moons</a></li>
</ul></li>
<li class="chapter" data-level="4.56" data-path="clustering.html"><a href="clustering.html#example-gene-expression-profiling-of-human-tissues"><i class="fa fa-check"></i><b>4.56</b> Example: gene expression profiling of human tissues</a>
<ul>
<li class="chapter" data-level="4.56.1" data-path="clustering.html"><a href="clustering.html#hierarchic-agglomerative-2"><i class="fa fa-check"></i><b>4.56.1</b> Hierarchic agglomerative</a></li>
</ul></li>
<li class="chapter" data-level="4.57" data-path="clustering.html"><a href="clustering.html#tissue"><i class="fa fa-check"></i><b>4.57</b> tissue</a></li>
<li class="chapter" data-level="4.58" data-path="clustering.html"><a href="clustering.html#cerebellum-colon-endometrium-hippocampus-kidney-liver"><i class="fa fa-check"></i><b>4.58</b> cerebellum colon endometrium hippocampus kidney liver</a></li>
<li class="chapter" data-level="4.59" data-path="clustering.html"><a href="clustering.html#section-30"><i class="fa fa-check"></i><b>4.59</b> 38 34 15 31 39 26</a></li>
<li class="chapter" data-level="4.60" data-path="clustering.html"><a href="clustering.html#placenta"><i class="fa fa-check"></i><b>4.60</b> placenta</a></li>
<li class="chapter" data-level="4.61" data-path="clustering.html"><a href="clustering.html#section-31"><i class="fa fa-check"></i><b>4.61</b> 6</a></li>
<li class="chapter" data-level="4.62" data-path="clustering.html"><a href="clustering.html#section-32"><i class="fa fa-check"></i><b>4.62</b> <span>1</span> 22215 189</a></li>
<li class="chapter" data-level="4.63" data-path="clustering.html"><a href="clustering.html#cluster"><i class="fa fa-check"></i><b>4.63</b> cluster</a></li>
<li class="chapter" data-level="4.64" data-path="clustering.html"><a href="clustering.html#tissue-1-2-3-4-5-6"><i class="fa fa-check"></i><b>4.64</b> tissue 1 2 3 4 5 6</a></li>
<li class="chapter" data-level="4.65" data-path="clustering.html"><a href="clustering.html#cerebellum-0-36-0-0-2-0"><i class="fa fa-check"></i><b>4.65</b> cerebellum 0 36 0 0 2 0</a></li>
<li class="chapter" data-level="4.66" data-path="clustering.html"><a href="clustering.html#colon-0-0-34-0-0-0"><i class="fa fa-check"></i><b>4.66</b> colon 0 0 34 0 0 0</a></li>
<li class="chapter" data-level="4.67" data-path="clustering.html"><a href="clustering.html#endometrium-15-0-0-0-0-0"><i class="fa fa-check"></i><b>4.67</b> endometrium 15 0 0 0 0 0</a></li>
<li class="chapter" data-level="4.68" data-path="clustering.html"><a href="clustering.html#hippocampus-0-31-0-0-0-0"><i class="fa fa-check"></i><b>4.68</b> hippocampus 0 31 0 0 0 0</a></li>
<li class="chapter" data-level="4.69" data-path="clustering.html"><a href="clustering.html#kidney-37-0-0-0-2-0"><i class="fa fa-check"></i><b>4.69</b> kidney 37 0 0 0 2 0</a></li>
<li class="chapter" data-level="4.70" data-path="clustering.html"><a href="clustering.html#liver-0-0-0-24-2-0"><i class="fa fa-check"></i><b>4.70</b> liver 0 0 0 24 2 0</a></li>
<li class="chapter" data-level="4.71" data-path="clustering.html"><a href="clustering.html#placenta-0-0-0-0-0-6"><i class="fa fa-check"></i><b>4.71</b> placenta 0 0 0 0 0 6</a></li>
<li class="chapter" data-level="4.72" data-path="clustering.html"><a href="clustering.html#cluster-1"><i class="fa fa-check"></i><b>4.72</b> cluster</a></li>
<li class="chapter" data-level="4.73" data-path="clustering.html"><a href="clustering.html#tissue-1-2-3-4-5-6-7-8"><i class="fa fa-check"></i><b>4.73</b> tissue 1 2 3 4 5 6 7 8</a></li>
<li class="chapter" data-level="4.74" data-path="clustering.html"><a href="clustering.html#cerebellum-0-31-0-0-2-0-5-0"><i class="fa fa-check"></i><b>4.74</b> cerebellum 0 31 0 0 2 0 5 0</a></li>
<li class="chapter" data-level="4.75" data-path="clustering.html"><a href="clustering.html#colon-0-0-34-0-0-0-0-0"><i class="fa fa-check"></i><b>4.75</b> colon 0 0 34 0 0 0 0 0</a></li>
<li class="chapter" data-level="4.76" data-path="clustering.html"><a href="clustering.html#endometrium-0-0-0-0-0-15-0-0"><i class="fa fa-check"></i><b>4.76</b> endometrium 0 0 0 0 0 15 0 0</a></li>
<li class="chapter" data-level="4.77" data-path="clustering.html"><a href="clustering.html#hippocampus-0-31-0-0-0-0-0-0"><i class="fa fa-check"></i><b>4.77</b> hippocampus 0 31 0 0 0 0 0 0</a></li>
<li class="chapter" data-level="4.78" data-path="clustering.html"><a href="clustering.html#kidney-37-0-0-0-2-0-0-0"><i class="fa fa-check"></i><b>4.78</b> kidney 37 0 0 0 2 0 0 0</a></li>
<li class="chapter" data-level="4.79" data-path="clustering.html"><a href="clustering.html#liver-0-0-0-24-2-0-0-0"><i class="fa fa-check"></i><b>4.79</b> liver 0 0 0 24 2 0 0 0</a></li>
<li class="chapter" data-level="4.80" data-path="clustering.html"><a href="clustering.html#placenta-0-0-0-0-0-0-0-6"><i class="fa fa-check"></i><b>4.80</b> placenta 0 0 0 0 0 0 0 6</a></li>
<li class="chapter" data-level="4.81" data-path="clustering.html"><a href="clustering.html#section-33"><i class="fa fa-check"></i><b>4.81</b> </a></li>
<li class="chapter" data-level="4.82" data-path="clustering.html"><a href="clustering.html#attaching-package-gplots"><i class="fa fa-check"></i><b>4.82</b> Attaching package: ‘gplots’</a></li>
<li class="chapter" data-level="4.83" data-path="clustering.html"><a href="clustering.html#the-following-object-is-masked-from-packagestats-1"><i class="fa fa-check"></i><b>4.83</b> The following object is masked from ‘package:stats’:</a></li>
<li class="chapter" data-level="4.84" data-path="clustering.html"><a href="clustering.html#section-34"><i class="fa fa-check"></i><b>4.84</b> </a></li>
<li class="chapter" data-level="4.85" data-path="clustering.html"><a href="clustering.html#lowess"><i class="fa fa-check"></i><b>4.85</b> lowess</a>
<ul>
<li class="chapter" data-level="4.85.1" data-path="clustering.html"><a href="clustering.html#k-means-2"><i class="fa fa-check"></i><b>4.85.1</b> K-means</a></li>
</ul></li>
<li class="chapter" data-level="4.86" data-path="clustering.html"><a href="clustering.html#tissue-1"><i class="fa fa-check"></i><b>4.86</b> tissue</a></li>
<li class="chapter" data-level="4.87" data-path="clustering.html"><a href="clustering.html#cerebellum-colon-endometrium-hippocampus-kidney-liver-1"><i class="fa fa-check"></i><b>4.87</b> cerebellum colon endometrium hippocampus kidney liver</a></li>
<li class="chapter" data-level="4.88" data-path="clustering.html"><a href="clustering.html#section-35"><i class="fa fa-check"></i><b>4.88</b> 38 34 15 31 39 26</a></li>
<li class="chapter" data-level="4.89" data-path="clustering.html"><a href="clustering.html#placenta-1"><i class="fa fa-check"></i><b>4.89</b> placenta</a></li>
<li class="chapter" data-level="4.90" data-path="clustering.html"><a href="clustering.html#section-36"><i class="fa fa-check"></i><b>4.90</b> 6</a></li>
<li class="chapter" data-level="4.91" data-path="clustering.html"><a href="clustering.html#section-37"><i class="fa fa-check"></i><b>4.91</b> <span>1</span> 22215 189</a></li>
<li class="chapter" data-level="4.92" data-path="clustering.html"><a href="clustering.html#loading-required-package-foreach"><i class="fa fa-check"></i><b>4.92</b> Loading required package: foreach</a></li>
<li class="chapter" data-level="4.93" data-path="clustering.html"><a href="clustering.html#loading-required-package-iterators"><i class="fa fa-check"></i><b>4.93</b> Loading required package: iterators</a></li>
<li class="chapter" data-level="4.94" data-path="clustering.html"><a href="clustering.html#loading-required-package-parallel"><i class="fa fa-check"></i><b>4.94</b> Loading required package: parallel</a></li>
<li class="chapter" data-level="4.95" data-path="clustering.html"><a href="clustering.html#section-38"><i class="fa fa-check"></i><b>4.95</b> <span>1</span> 8</a></li>
<li class="chapter" data-level="4.96" data-path="clustering.html"><a href="clustering.html#section-39"><i class="fa fa-check"></i><b>4.96</b> </a></li>
<li class="chapter" data-level="4.97" data-path="clustering.html"><a href="clustering.html#tissue-1-2-3-4-5-6-7"><i class="fa fa-check"></i><b>4.97</b> tissue 1 2 3 4 5 6 7</a></li>
<li class="chapter" data-level="4.98" data-path="clustering.html"><a href="clustering.html#cerebellum-33-5-0-0-0-0-0"><i class="fa fa-check"></i><b>4.98</b> cerebellum 33 5 0 0 0 0 0</a></li>
<li class="chapter" data-level="4.99" data-path="clustering.html"><a href="clustering.html#colon-0-0-0-0-0-34-0"><i class="fa fa-check"></i><b>4.99</b> colon 0 0 0 0 0 34 0</a></li>
<li class="chapter" data-level="4.100" data-path="clustering.html"><a href="clustering.html#endometrium-0-0-15-0-0-0-0"><i class="fa fa-check"></i><b>4.100</b> endometrium 0 0 15 0 0 0 0</a></li>
<li class="chapter" data-level="4.101" data-path="clustering.html"><a href="clustering.html#hippocampus-0-31-0-0-0-0-0"><i class="fa fa-check"></i><b>4.101</b> hippocampus 0 31 0 0 0 0 0</a></li>
<li class="chapter" data-level="4.102" data-path="clustering.html"><a href="clustering.html#kidney-0-0-0-39-0-0-0"><i class="fa fa-check"></i><b>4.102</b> kidney 0 0 0 39 0 0 0</a></li>
<li class="chapter" data-level="4.103" data-path="clustering.html"><a href="clustering.html#liver-0-0-0-0-26-0-0"><i class="fa fa-check"></i><b>4.103</b> liver 0 0 0 0 26 0 0</a></li>
<li class="chapter" data-level="4.104" data-path="clustering.html"><a href="clustering.html#placenta-0-0-0-0-0-0-6"><i class="fa fa-check"></i><b>4.104</b> placenta 0 0 0 0 0 0 6</a></li>
<li class="chapter" data-level="4.105" data-path="clustering.html"><a href="clustering.html#section-40"><i class="fa fa-check"></i><b>4.105</b> </a></li>
<li class="chapter" data-level="4.106" data-path="clustering.html"><a href="clustering.html#tissue-1-2-3-4-5-6-7-1"><i class="fa fa-check"></i><b>4.106</b> tissue 1 2 3 4 5 6 7</a></li>
<li class="chapter" data-level="4.107" data-path="clustering.html"><a href="clustering.html#cerebellum-33-5-0-0-0-0-0-1"><i class="fa fa-check"></i><b>4.107</b> cerebellum 33 5 0 0 0 0 0</a></li>
<li class="chapter" data-level="4.108" data-path="clustering.html"><a href="clustering.html#colon-0-0-0-0-0-34-0-1"><i class="fa fa-check"></i><b>4.108</b> colon 0 0 0 0 0 34 0</a></li>
<li class="chapter" data-level="4.109" data-path="clustering.html"><a href="clustering.html#endometrium-0-0-15-0-0-0-0-1"><i class="fa fa-check"></i><b>4.109</b> endometrium 0 0 15 0 0 0 0</a></li>
<li class="chapter" data-level="4.110" data-path="clustering.html"><a href="clustering.html#hippocampus-0-31-0-0-0-0-0-1"><i class="fa fa-check"></i><b>4.110</b> hippocampus 0 31 0 0 0 0 0</a></li>
<li class="chapter" data-level="4.111" data-path="clustering.html"><a href="clustering.html#kidney-0-0-0-39-0-0-0-1"><i class="fa fa-check"></i><b>4.111</b> kidney 0 0 0 39 0 0 0</a></li>
<li class="chapter" data-level="4.112" data-path="clustering.html"><a href="clustering.html#liver-0-0-0-0-26-0-0-1"><i class="fa fa-check"></i><b>4.112</b> liver 0 0 0 0 26 0 0</a></li>
<li class="chapter" data-level="4.113" data-path="clustering.html"><a href="clustering.html#placenta-0-0-0-0-0-0-6-1"><i class="fa fa-check"></i><b>4.113</b> placenta 0 0 0 0 0 0 6</a>
<ul>
<li class="chapter" data-level="4.113.1" data-path="clustering.html"><a href="clustering.html#dbscan-2"><i class="fa fa-check"></i><b>4.113.1</b> DBSCAN</a></li>
</ul></li>
<li class="chapter" data-level="4.114" data-path="clustering.html"><a href="clustering.html#tissue-2"><i class="fa fa-check"></i><b>4.114</b> tissue</a></li>
<li class="chapter" data-level="4.115" data-path="clustering.html"><a href="clustering.html#cerebellum-colon-endometrium-hippocampus-kidney-liver-2"><i class="fa fa-check"></i><b>4.115</b> cerebellum colon endometrium hippocampus kidney liver</a></li>
<li class="chapter" data-level="4.116" data-path="clustering.html"><a href="clustering.html#section-41"><i class="fa fa-check"></i><b>4.116</b> 38 34 15 31 39 26</a></li>
<li class="chapter" data-level="4.117" data-path="clustering.html"><a href="clustering.html#placenta-2"><i class="fa fa-check"></i><b>4.117</b> placenta</a></li>
<li class="chapter" data-level="4.118" data-path="clustering.html"><a href="clustering.html#section-42"><i class="fa fa-check"></i><b>4.118</b> 6</a></li>
<li class="chapter" data-level="4.119" data-path="clustering.html"><a href="clustering.html#section-43"><i class="fa fa-check"></i><b>4.119</b> </a></li>
<li class="chapter" data-level="4.120" data-path="clustering.html"><a href="clustering.html#section-44"><i class="fa fa-check"></i><b>4.120</b> 0 1 2 3 4 5 6</a></li>
<li class="chapter" data-level="4.121" data-path="clustering.html"><a href="clustering.html#section-45"><i class="fa fa-check"></i><b>4.121</b> 12 37 62 34 24 15 5</a></li>
<li class="chapter" data-level="4.122" data-path="clustering.html"><a href="clustering.html#section-46"><i class="fa fa-check"></i><b>4.122</b> </a></li>
<li class="chapter" data-level="4.123" data-path="clustering.html"><a href="clustering.html#tissue-0-1-2-3-4-5-6"><i class="fa fa-check"></i><b>4.123</b> tissue 0 1 2 3 4 5 6</a></li>
<li class="chapter" data-level="4.124" data-path="clustering.html"><a href="clustering.html#cerebellum-2-0-31-0-0-0-5"><i class="fa fa-check"></i><b>4.124</b> cerebellum 2 0 31 0 0 0 5</a></li>
<li class="chapter" data-level="4.125" data-path="clustering.html"><a href="clustering.html#colon-0-0-0-34-0-0-0"><i class="fa fa-check"></i><b>4.125</b> colon 0 0 0 34 0 0 0</a></li>
<li class="chapter" data-level="4.126" data-path="clustering.html"><a href="clustering.html#endometrium-0-0-0-0-0-15-0"><i class="fa fa-check"></i><b>4.126</b> endometrium 0 0 0 0 0 15 0</a></li>
<li class="chapter" data-level="4.127" data-path="clustering.html"><a href="clustering.html#hippocampus-0-0-31-0-0-0-0"><i class="fa fa-check"></i><b>4.127</b> hippocampus 0 0 31 0 0 0 0</a></li>
<li class="chapter" data-level="4.128" data-path="clustering.html"><a href="clustering.html#kidney-2-37-0-0-0-0-0"><i class="fa fa-check"></i><b>4.128</b> kidney 2 37 0 0 0 0 0</a></li>
<li class="chapter" data-level="4.129" data-path="clustering.html"><a href="clustering.html#liver-2-0-0-0-24-0-0"><i class="fa fa-check"></i><b>4.129</b> liver 2 0 0 0 24 0 0</a></li>
<li class="chapter" data-level="4.130" data-path="clustering.html"><a href="clustering.html#placenta-6-0-0-0-0-0-0"><i class="fa fa-check"></i><b>4.130</b> placenta 6 0 0 0 0 0 0</a></li>
<li class="chapter" data-level="4.131" data-path="clustering.html"><a href="clustering.html#exercises"><i class="fa fa-check"></i><b>4.131</b> Exercises</a>
<ul>
<li class="chapter" data-level="4.131.1" data-path="clustering.html"><a href="clustering.html#clusteringEx1"><i class="fa fa-check"></i><b>4.131.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="4.132" data-path="clustering.html"><a href="clustering.html#section-47"><i class="fa fa-check"></i><b>4.132</b> </a></li>
<li class="chapter" data-level="4.133" data-path="clustering.html"><a href="clustering.html#attaching-package-ebimage"><i class="fa fa-check"></i><b>4.133</b> Attaching package: ‘EBImage’</a></li>
<li class="chapter" data-level="4.134" data-path="clustering.html"><a href="clustering.html#the-following-object-is-masked-from-packagedendextend"><i class="fa fa-check"></i><b>4.134</b> The following object is masked from ‘package:dendextend’:</a></li>
<li class="chapter" data-level="4.135" data-path="clustering.html"><a href="clustering.html#section-48"><i class="fa fa-check"></i><b>4.135</b> </a></li>
<li class="chapter" data-level="4.136" data-path="clustering.html"><a href="clustering.html#rotate"><i class="fa fa-check"></i><b>4.136</b> rotate</a></li>
<li class="chapter" data-level="4.137" data-path="clustering.html"><a href="clustering.html#section-49"><i class="fa fa-check"></i><b>4.137</b> <span>1</span> 528 393 3</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html"><i class="fa fa-check"></i><b>5</b> Nearest neighbours</a>
<ul>
<li class="chapter" data-level="5.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#introduction-1"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#classification-simulated-data"><i class="fa fa-check"></i><b>5.2</b> Classification: simulated data</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#knn-function"><i class="fa fa-check"></i><b>5.2.1</b> knn function</a></li>
<li class="chapter" data-level="5.2.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#plotting-decision-boundaries"><i class="fa fa-check"></i><b>5.2.2</b> Plotting decision boundaries</a></li>
<li class="chapter" data-level="5.2.3" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#bias-variance-tradeoff"><i class="fa fa-check"></i><b>5.2.3</b> Bias-variance tradeoff</a></li>
<li class="chapter" data-level="5.2.4" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#choosing-k"><i class="fa fa-check"></i><b>5.2.4</b> Choosing <em>k</em></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#example-on-the-iris-dataset"><i class="fa fa-check"></i><b>5.3</b> Example on the Iris dataset</a></li>
<li class="chapter" data-level="5.4" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#knn-cell-segmentation"><i class="fa fa-check"></i><b>5.4</b> Classification: cell segmentation</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#cell-segmentation-data-set"><i class="fa fa-check"></i><b>5.4.1</b> Cell segmentation data set</a></li>
<li class="chapter" data-level="5.4.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#data-splitting"><i class="fa fa-check"></i><b>5.4.2</b> Data splitting</a></li>
<li class="chapter" data-level="5.4.3" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#identification-of-data-quality-issues"><i class="fa fa-check"></i><b>5.4.3</b> Identification of data quality issues</a></li>
<li class="chapter" data-level="5.4.4" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#fit-model"><i class="fa fa-check"></i><b>5.4.4</b> Fit model</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#knn-regression"><i class="fa fa-check"></i><b>5.5</b> Regression</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#partition-data"><i class="fa fa-check"></i><b>5.5.1</b> Partition data</a></li>
<li class="chapter" data-level="5.5.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#data-pre-processing"><i class="fa fa-check"></i><b>5.5.2</b> Data pre-processing</a></li>
<li class="chapter" data-level="5.5.3" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#search-for-optimum-k"><i class="fa fa-check"></i><b>5.5.3</b> Search for optimum <em>k</em></a></li>
<li class="chapter" data-level="5.5.4" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#use-model-to-make-predictions"><i class="fa fa-check"></i><b>5.5.4</b> Use model to make predictions</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#exercises-1"><i class="fa fa-check"></i><b>5.6</b> Exercises</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#knnEx1"><i class="fa fa-check"></i><b>5.6.1</b> Exercise 1</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="svm.html"><a href="svm.html"><i class="fa fa-check"></i><b>6</b> Support vector machines</a>
<ul>
<li class="chapter" data-level="6.1" data-path="svm.html"><a href="svm.html#introduction-2"><i class="fa fa-check"></i><b>6.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="svm.html"><a href="svm.html#maximum-margin-classifier"><i class="fa fa-check"></i><b>6.1.1</b> Maximum margin classifier</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="svm.html"><a href="svm.html#support-vector-classifier"><i class="fa fa-check"></i><b>6.2</b> Support vector classifier</a></li>
<li class="chapter" data-level="6.3" data-path="svm.html"><a href="svm.html#support-vector-machine"><i class="fa fa-check"></i><b>6.3</b> Support Vector Machine</a></li>
<li class="chapter" data-level="6.4" data-path="svm.html"><a href="svm.html#example---training-a-classifier"><i class="fa fa-check"></i><b>6.4</b> Example - training a classifier</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="svm.html"><a href="svm.html#setup-environment"><i class="fa fa-check"></i><b>6.4.1</b> Setup environment</a></li>
<li class="chapter" data-level="6.4.2" data-path="svm.html"><a href="svm.html#partition-data-1"><i class="fa fa-check"></i><b>6.4.2</b> Partition data</a></li>
<li class="chapter" data-level="6.4.3" data-path="svm.html"><a href="svm.html#visualize-training-data"><i class="fa fa-check"></i><b>6.4.3</b> Visualize training data</a></li>
<li class="chapter" data-level="6.4.4" data-path="svm.html"><a href="svm.html#prediction-performance-measures"><i class="fa fa-check"></i><b>6.4.4</b> Prediction performance measures</a></li>
<li class="chapter" data-level="6.4.5" data-path="svm.html"><a href="svm.html#plot-decision-boundary"><i class="fa fa-check"></i><b>6.4.5</b> Plot decision boundary</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="svm.html"><a href="svm.html#defining-your-own-model-type-to-use-in-caret"><i class="fa fa-check"></i><b>6.5</b> Defining your own model type to use in caret</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="svm.html"><a href="svm.html#model-cross-validation-and-tuning"><i class="fa fa-check"></i><b>6.5.1</b> Model cross-validation and tuning</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="svm.html"><a href="svm.html#iris-example"><i class="fa fa-check"></i><b>6.6</b> Iris example</a></li>
<li class="chapter" data-level="6.7" data-path="svm.html"><a href="svm.html#cell-segmentation-example"><i class="fa fa-check"></i><b>6.7</b> Cell segmentation example</a></li>
<li class="chapter" data-level="6.8" data-path="svm.html"><a href="svm.html#example---regression"><i class="fa fa-check"></i><b>6.8</b> Example - regression</a></li>
<li class="chapter" data-level="6.9" data-path="svm.html"><a href="svm.html#further-reading"><i class="fa fa-check"></i><b>6.9</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="decision-trees.html"><a href="decision-trees.html"><i class="fa fa-check"></i><b>7</b> Decision trees and random forests</a>
<ul>
<li class="chapter" data-level="7.1" data-path="decision-trees.html"><a href="decision-trees.html#decision-trees-1"><i class="fa fa-check"></i><b>7.1</b> Decision Trees</a></li>
<li class="chapter" data-level="7.2" data-path="decision-trees.html"><a href="decision-trees.html#example-code-with-categorical-data"><i class="fa fa-check"></i><b>7.2</b> Example code with categorical data</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="decision-trees.html"><a href="decision-trees.html#loading-packages-and-data"><i class="fa fa-check"></i><b>7.2.1</b> Loading packages and data</a></li>
<li class="chapter" data-level="7.2.2" data-path="decision-trees.html"><a href="decision-trees.html#data-slicing"><i class="fa fa-check"></i><b>7.2.2</b> Data Slicing</a></li>
<li class="chapter" data-level="7.2.3" data-path="decision-trees.html"><a href="decision-trees.html#data-preprocessing"><i class="fa fa-check"></i><b>7.2.3</b> Data Preprocessing</a></li>
<li class="chapter" data-level="7.2.4" data-path="decision-trees.html"><a href="decision-trees.html#training-decisions-trees"><i class="fa fa-check"></i><b>7.2.4</b> Training Decisions Trees</a></li>
<li class="chapter" data-level="7.2.5" data-path="decision-trees.html"><a href="decision-trees.html#plotting-the-decision-trees"><i class="fa fa-check"></i><b>7.2.5</b> Plotting the decision trees</a></li>
<li class="chapter" data-level="7.2.6" data-path="decision-trees.html"><a href="decision-trees.html#prediction"><i class="fa fa-check"></i><b>7.2.6</b> Prediction</a></li>
<li class="chapter" data-level="7.2.7" data-path="decision-trees.html"><a href="decision-trees.html#implementing-decision-trees-directly"><i class="fa fa-check"></i><b>7.2.7</b> Implementing Decision Trees directly</a></li>
<li class="chapter" data-level="7.2.8" data-path="decision-trees.html"><a href="decision-trees.html#iris-example-for-decision-trees"><i class="fa fa-check"></i><b>7.2.8</b> Iris example for Decision Trees</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="decision-trees.html"><a href="decision-trees.html#cell-segmentation-examples"><i class="fa fa-check"></i><b>7.3</b> Cell segmentation examples</a></li>
<li class="chapter" data-level="7.4" data-path="decision-trees.html"><a href="decision-trees.html#regression-example---blood-brain-barrier"><i class="fa fa-check"></i><b>7.4</b> Regression example - Blood Brain Barrier</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="decision-trees.html"><a href="decision-trees.html#decision-tree"><i class="fa fa-check"></i><b>7.4.1</b> Decision tree</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="decision-trees.html"><a href="decision-trees.html#random-forest"><i class="fa fa-check"></i><b>7.5</b> Random Forest</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="decision-trees.html"><a href="decision-trees.html#iris-example-for-random-forests"><i class="fa fa-check"></i><b>7.5.1</b> Iris example for Random Forests</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="decision-trees.html"><a href="decision-trees.html#cell-segmentation-examples-1"><i class="fa fa-check"></i><b>7.6</b> Cell segmentation examples</a></li>
<li class="chapter" data-level="7.7" data-path="decision-trees.html"><a href="decision-trees.html#regression-example---blood-brain-barrier-1"><i class="fa fa-check"></i><b>7.7</b> Regression example - Blood Brain Barrier</a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="decision-trees.html"><a href="decision-trees.html#random-forest-1"><i class="fa fa-check"></i><b>7.7.1</b> Random forest</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="use-case-1.html"><a href="use-case-1.html"><i class="fa fa-check"></i><b>8</b> Use case 1</a>
<ul>
<li class="chapter" data-level="8.1" data-path="use-case-1.html"><a href="use-case-1.html#introduction-3"><i class="fa fa-check"></i><b>8.1</b> Introduction</a></li>
<li class="chapter" data-level="8.2" data-path="use-case-1.html"><a href="use-case-1.html#problem-automated-detection-of-malaria"><i class="fa fa-check"></i><b>8.2</b> Problem: automated detection of malaria</a></li>
<li class="chapter" data-level="8.3" data-path="use-case-1.html"><a href="use-case-1.html#challenges"><i class="fa fa-check"></i><b>8.3</b> Challenges</a></li>
<li class="chapter" data-level="8.4" data-path="use-case-1.html"><a href="use-case-1.html#getting-started"><i class="fa fa-check"></i><b>8.4</b> Getting started</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="use-case-1.html"><a href="use-case-1.html#load-data"><i class="fa fa-check"></i><b>8.4.1</b> Load data</a></li>
<li class="chapter" data-level="8.4.2" data-path="use-case-1.html"><a href="use-case-1.html#model-comparison"><i class="fa fa-check"></i><b>8.4.2</b> Model comparison</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="use-case-1.html"><a href="use-case-1.html#solutions"><i class="fa fa-check"></i><b>8.5</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="linear-models.html"><a href="linear-models.html"><i class="fa fa-check"></i><b>9</b> Linear models and matrix algebra</a>
<ul>
<li class="chapter" data-level="9.1" data-path="linear-models.html"><a href="linear-models.html#linear-models-1"><i class="fa fa-check"></i><b>9.1</b> Linear models</a></li>
<li class="chapter" data-level="9.2" data-path="linear-models.html"><a href="linear-models.html#matrix-algebra"><i class="fa fa-check"></i><b>9.2</b> Matrix algebra</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="logistic-regression.html"><a href="logistic-regression.html"><i class="fa fa-check"></i><b>10</b> Linear regression and logistic regression</a>
<ul>
<li class="chapter" data-level="10.1" data-path="logistic-regression.html"><a href="logistic-regression.html#regression"><i class="fa fa-check"></i><b>10.1</b> Regression</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="logistic-regression.html"><a href="logistic-regression.html#linear-regression"><i class="fa fa-check"></i><b>10.1.1</b> Linear regression</a></li>
<li class="chapter" data-level="10.1.2" data-path="logistic-regression.html"><a href="logistic-regression.html#polynomial-regression"><i class="fa fa-check"></i><b>10.1.2</b> Polynomial regression</a></li>
<li class="chapter" data-level="10.1.3" data-path="logistic-regression.html"><a href="logistic-regression.html#distributions-of-fits"><i class="fa fa-check"></i><b>10.1.3</b> Distributions of fits</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="logistic-regression.html"><a href="logistic-regression.html#classification"><i class="fa fa-check"></i><b>10.2</b> Classification</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="logistic-regression.html"><a href="logistic-regression.html#logistic-regression"><i class="fa fa-check"></i><b>10.2.1</b> Logistic regression</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="logistic-regression.html"><a href="logistic-regression.html#resources"><i class="fa fa-check"></i><b>10.3</b> Resources</a></li>
<li class="chapter" data-level="10.4" data-path="logistic-regression.html"><a href="logistic-regression.html#exercises-2"><i class="fa fa-check"></i><b>10.4</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="ann.html"><a href="ann.html"><i class="fa fa-check"></i><b>11</b> Artificial neural networks</a>
<ul>
<li class="chapter" data-level="11.1" data-path="ann.html"><a href="ann.html#neural-networks"><i class="fa fa-check"></i><b>11.1</b> Neural Networks</a></li>
<li class="chapter" data-level="11.2" data-path="ann.html"><a href="ann.html#exercises-3"><i class="fa fa-check"></i><b>11.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="use-case-2.html"><a href="use-case-2.html"><i class="fa fa-check"></i><b>12</b> Use case 2</a></li>
<li class="chapter" data-level="13" data-path="mlnn.html"><a href="mlnn.html"><i class="fa fa-check"></i><b>13</b> Deep Learning</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mlnn.html"><a href="mlnn.html#multilayer-neural-networks"><i class="fa fa-check"></i><b>13.1</b> Multilayer Neural Networks</a>
<ul>
<li class="chapter" data-level="13.1.1" data-path="mlnn.html"><a href="mlnn.html#reading-in-images"><i class="fa fa-check"></i><b>13.1.1</b> Reading in images</a></li>
<li class="chapter" data-level="13.1.2" data-path="mlnn.html"><a href="mlnn.html#constructing-layers-in-kerasr"><i class="fa fa-check"></i><b>13.1.2</b> Constructing layers in kerasR</a></li>
<li class="chapter" data-level="13.1.3" data-path="mlnn.html"><a href="mlnn.html#rick-and-morty-classifier-using-deep-learning"><i class="fa fa-check"></i><b>13.1.3</b> Rick and Morty classifier using Deep Learning</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="mlnn.html"><a href="mlnn.html#convolutional-neural-networks"><i class="fa fa-check"></i><b>13.2</b> Convolutional neural networks</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="mlnn.html"><a href="mlnn.html#checking-the-models"><i class="fa fa-check"></i><b>13.2.1</b> Checking the models</a></li>
<li class="chapter" data-level="13.2.2" data-path="mlnn.html"><a href="mlnn.html#data-augmentation"><i class="fa fa-check"></i><b>13.2.2</b> Data augmentation</a></li>
<li class="chapter" data-level="13.2.3" data-path="mlnn.html"><a href="mlnn.html#asking-more-precise-questions"><i class="fa fa-check"></i><b>13.2.3</b> Asking more precise questions</a></li>
<li class="chapter" data-level="13.2.4" data-path="mlnn.html"><a href="mlnn.html#more-complex-networks"><i class="fa fa-check"></i><b>13.2.4</b> More complex networks</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="mlnn.html"><a href="mlnn.html#further-reading-1"><i class="fa fa-check"></i><b>13.3</b> Further reading</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="logistic-regression.html"><a href="logistic-regression.html#resources"><i class="fa fa-check"></i><b>A</b> Resources</a>
<ul>
<li class="chapter" data-level="A.1" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>A.1</b> Python</a></li>
<li class="chapter" data-level="A.2" data-path="resources.html"><a href="resources.html#machine-learning-data-set-repositories"><i class="fa fa-check"></i><b>A.2</b> Machine learning data set repositories</a>
<ul>
<li class="chapter" data-level="A.2.1" data-path="resources.html"><a href="resources.html#mldata"><i class="fa fa-check"></i><b>A.2.1</b> MLDATA</a></li>
<li class="chapter" data-level="A.2.2" data-path="resources.html"><a href="resources.html#uci-machine-learning-repository"><i class="fa fa-check"></i><b>A.2.2</b> UCI Machine Learning Repository</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="B" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html"><i class="fa fa-check"></i><b>B</b> Solutions ch. 2 - Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="B.1" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-2.5."><i class="fa fa-check"></i><b>B.1</b> Exercise 2.5.</a></li>
<li class="chapter" data-level="B.2" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-2.6."><i class="fa fa-check"></i><b>B.2</b> Exercise 2.6.</a></li>
</ul></li>
<li class="chapter" data-level="C" data-path="solutions-clustering.html"><a href="solutions-clustering.html"><i class="fa fa-check"></i><b>C</b> Solutions ch. 4 - Clustering</a>
<ul>
<li class="chapter" data-level="C.1" data-path="solutions-clustering.html"><a href="solutions-clustering.html#exercise-1"><i class="fa fa-check"></i><b>C.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="D" data-path="solutions-nearest-neighbours.html"><a href="solutions-nearest-neighbours.html"><i class="fa fa-check"></i><b>D</b> Solutions ch. 7 - Nearest neighbours</a>
<ul>
<li class="chapter" data-level="D.1" data-path="solutions-nearest-neighbours.html"><a href="solutions-nearest-neighbours.html#exercise-1-1"><i class="fa fa-check"></i><b>D.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="E" data-path="solutions-svm.html"><a href="solutions-svm.html"><i class="fa fa-check"></i><b>E</b> Solutions ch. 6 - Support vector machines</a>
<ul>
<li class="chapter" data-level="E.1" data-path="solutions-svm.html"><a href="solutions-svm.html#exercise-1-2"><i class="fa fa-check"></i><b>E.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="F" data-path="solutions-decision-trees.html"><a href="solutions-decision-trees.html"><i class="fa fa-check"></i><b>F</b> Solutions ch. 9 - Decision trees and random forests</a>
<ul>
<li class="chapter" data-level="F.1" data-path="solutions-decision-trees.html"><a href="solutions-decision-trees.html#exercise-1-3"><i class="fa fa-check"></i><b>F.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="G" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html"><i class="fa fa-check"></i><b>G</b> Solutions chapter 8 - use case 1</a>
<ul>
<li class="chapter" data-level="G.1" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#preparation"><i class="fa fa-check"></i><b>G.1</b> Preparation</a>
<ul>
<li class="chapter" data-level="G.1.1" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#load-required-libraries"><i class="fa fa-check"></i><b>G.1.1</b> Load required libraries</a></li>
<li class="chapter" data-level="G.1.2" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#define-svm-model"><i class="fa fa-check"></i><b>G.1.2</b> Define SVM model</a></li>
<li class="chapter" data-level="G.1.3" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#setup-parallel-processing"><i class="fa fa-check"></i><b>G.1.3</b> Setup parallel processing</a></li>
<li class="chapter" data-level="G.1.4" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#load-data-1"><i class="fa fa-check"></i><b>G.1.4</b> Load data</a></li>
</ul></li>
<li class="chapter" data-level="G.2" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#assess-data-quality"><i class="fa fa-check"></i><b>G.2</b> Assess data quality</a>
<ul>
<li class="chapter" data-level="G.2.1" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#zero-and-near-zero-variance-predictors-1"><i class="fa fa-check"></i><b>G.2.1</b> Zero and near-zero variance predictors</a></li>
<li class="chapter" data-level="G.2.2" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#are-all-predictors-on-the-same-scale"><i class="fa fa-check"></i><b>G.2.2</b> Are all predictors on the same scale?</a></li>
<li class="chapter" data-level="G.2.3" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#redundancy-from-correlated-variables"><i class="fa fa-check"></i><b>G.2.3</b> Redundancy from correlated variables</a></li>
<li class="chapter" data-level="G.2.4" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#skewness-1"><i class="fa fa-check"></i><b>G.2.4</b> Skewness</a></li>
</ul></li>
<li class="chapter" data-level="G.3" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#infection-status-two-class-problem"><i class="fa fa-check"></i><b>G.3</b> Infection status (two-class problem)</a>
<ul>
<li class="chapter" data-level="G.3.1" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#model-training-and-parameter-tuning"><i class="fa fa-check"></i><b>G.3.1</b> Model training and parameter tuning</a></li>
<li class="chapter" data-level="G.3.2" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#knn"><i class="fa fa-check"></i><b>G.3.2</b> KNN</a></li>
<li class="chapter" data-level="G.3.3" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#svm-1"><i class="fa fa-check"></i><b>G.3.3</b> SVM</a></li>
<li class="chapter" data-level="G.3.4" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#decision-tree-1"><i class="fa fa-check"></i><b>G.3.4</b> Decision tree</a></li>
<li class="chapter" data-level="G.3.5" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#random-forest-2"><i class="fa fa-check"></i><b>G.3.5</b> Random forest</a></li>
<li class="chapter" data-level="G.3.6" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#compare-models"><i class="fa fa-check"></i><b>G.3.6</b> Compare models</a></li>
<li class="chapter" data-level="G.3.7" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#predict-test-set-using-our-best-model"><i class="fa fa-check"></i><b>G.3.7</b> Predict test set using our best model</a></li>
<li class="chapter" data-level="G.3.8" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#roc-curve"><i class="fa fa-check"></i><b>G.3.8</b> ROC curve</a></li>
</ul></li>
<li class="chapter" data-level="G.4" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#discrimination-of-infective-stages-multi-class-problem"><i class="fa fa-check"></i><b>G.4</b> Discrimination of infective stages (multi-class problem)</a>
<ul>
<li class="chapter" data-level="G.4.1" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#define-cross-validation-procedure"><i class="fa fa-check"></i><b>G.4.1</b> Define cross-validation procedure</a></li>
<li class="chapter" data-level="G.4.2" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#knn-1"><i class="fa fa-check"></i><b>G.4.2</b> KNN</a></li>
<li class="chapter" data-level="G.4.3" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#svm-2"><i class="fa fa-check"></i><b>G.4.3</b> SVM</a></li>
<li class="chapter" data-level="G.4.4" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#decision-tree-2"><i class="fa fa-check"></i><b>G.4.4</b> Decision tree</a></li>
<li class="chapter" data-level="G.4.5" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#random-forest-3"><i class="fa fa-check"></i><b>G.4.5</b> Random forest</a></li>
<li class="chapter" data-level="G.4.6" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#compare-models-1"><i class="fa fa-check"></i><b>G.4.6</b> Compare models</a></li>
<li class="chapter" data-level="G.4.7" data-path="use-case-1-solutions.html"><a href="use-case-1-solutions.html#predict-test-set-using-our-best-model-1"><i class="fa fa-check"></i><b>G.4.7</b> Predict test set using our best model</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="H" data-path="solutions-linear-models.html"><a href="solutions-linear-models.html"><i class="fa fa-check"></i><b>H</b> Solutions ch. 3 - Linear models and matrix algebra</a>
<ul>
<li class="chapter" data-level="H.1" data-path="solutions-linear-models.html"><a href="solutions-linear-models.html#example-2"><i class="fa fa-check"></i><b>H.1</b> Example 2</a></li>
<li class="chapter" data-level="H.2" data-path="solutions-linear-models.html"><a href="solutions-linear-models.html#example-2-1"><i class="fa fa-check"></i><b>H.2</b> Example 2</a></li>
</ul></li>
<li class="chapter" data-level="I" data-path="solutions-logistic-regression.html"><a href="solutions-logistic-regression.html"><i class="fa fa-check"></i><b>I</b> Solutions ch. 4 - Linear and non-linear (logistic) regression</a></li>
<li class="chapter" data-level="J" data-path="solutions-ann.html"><a href="solutions-ann.html"><i class="fa fa-check"></i><b>J</b> Solutions ch. 10 - Artificial neural networks</a>
<ul>
<li class="chapter" data-level="J.1" data-path="solutions-ann.html"><a href="solutions-ann.html#exercise-1-4"><i class="fa fa-check"></i><b>J.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="K" data-path="use-case-2-solutions.html"><a href="use-case-2-solutions.html"><i class="fa fa-check"></i><b>K</b> Solutions for use case 2</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">An Introduction to Machine Learning</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="svm" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">6</span> Support vector machines<a href="svm.html#svm" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="introduction-2" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> Introduction<a href="svm.html#introduction-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Support vector machines (SVMs) are models of supervised learning, applicable to both classification and regression problems. The SVM is an extension of the support vector classifier (SVC), which is turn is an extension of the maximum margin classifier.</p>
<div id="maximum-margin-classifier" class="section level3 hasAnchor" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> Maximum margin classifier<a href="svm.html#maximum-margin-classifier" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s start by definining a hyperplane. In <em>p</em>-dimensional space a hyperplane is a flat affine subspace of <em>p</em>-1. Figure <a href="svm.html#fig:svmSeparatingHyperplanes2">6.1</a> shows three separating hyperplanes and objects of two different classes. A separating hyperplane forms a natural linear decision boundary, classifying new objects according to which side of the line they are located.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmSeparatingHyperplanes2"></span>
<img src="images/svm.9.2.png" alt="Left: two classes of observations (blue, purple) and three separating hyperplanes. Right: separating hyperplane shown as black line and grid indicates decision rule. Source: http://www-bcf.usc.edu/~gareth/ISL/" width="90%" />
<p class="caption">
Figure 6.1: Left: two classes of observations (blue, purple) and three separating hyperplanes. Right: separating hyperplane shown as black line and grid indicates decision rule. Source: <a href="http://www-bcf.usc.edu/~gareth/ISL/" class="uri">http://www-bcf.usc.edu/~gareth/ISL/</a>
</p>
</div>
<p>If the classes of observations can be separated by a hyperplane, then there will in fact be an infinite number of hyperplanes. So which of the possible hyperplanes do we choose to be our decision boundary?</p>
<p>The <strong>maximal margin hyperplane</strong> is the separating hyperplane that is farthest from the training observations. The perpendicular distance from a given hyperplane to the nearest training observation is known as the <strong>margin</strong>. The maximal margin hyperplane is the separating hyperplane for which the margin is largest.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmMaximalMarginHyperplane"></span>
<img src="images/svm.9.3.png" alt="Maximal margin hyperplane shown as solid line. Margin is the distance from the solid line to either of the dashed lines. The support vectors are the points on the dashed line. Source: http://www-bcf.usc.edu/~gareth/ISL/" width="75%" />
<p class="caption">
Figure 6.2: Maximal margin hyperplane shown as solid line. Margin is the distance from the solid line to either of the dashed lines. The support vectors are the points on the dashed line. Source: <a href="http://www-bcf.usc.edu/~gareth/ISL/" class="uri">http://www-bcf.usc.edu/~gareth/ISL/</a>
</p>
</div>
<p>Figure <a href="svm.html#fig:svmMaximalMarginHyperplane">6.2</a> shows three training observations that are equidistant from the maximal margin hyperplane and lie on the dashed lines indicating the margin. These are the <strong>support vectors</strong>. If these points were moved slightly, the maximal margin hyperplane would also move, hence the term <em>support</em>. The maximal margin hyperplane is set by the <strong>support vectors</strong> alone; it is not influenced by any other observations.</p>
The maximal margin hyperplane is a natural decision boundary, but only if a separating hyperplane exists. In practice there may be non separable cases which prevent the use of the maximal margin classifier.
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmNonSeparableCase"></span>
<img src="images/svm.9.4.png" alt="The two classes cannot be separated by a hyperplane and so the maximal margin classifier cannot be used. Source: http://www-bcf.usc.edu/~gareth/ISL/" width="75%" />
<p class="caption">
Figure 6.3: The two classes cannot be separated by a hyperplane and so the maximal margin classifier cannot be used. Source: <a href="http://www-bcf.usc.edu/~gareth/ISL/" class="uri">http://www-bcf.usc.edu/~gareth/ISL/</a>
</p>
</div>
</div>
</div>
<div id="support-vector-classifier" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> Support vector classifier<a href="svm.html#support-vector-classifier" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Even if a separating hyperplane exists, it may not be the best decision boundary. The maximal margin classifier is extremely sensitive to individual observations, so may overfit the training data.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmHyperplaneShift"></span>
<img src="images/svm.9.5.png" alt="Left: two classes of observations and a maximum margin hyperplane (solid line). Right: Hyperplane (solid line) moves after the addition of a new observation (original hyperplane is dashed line). Source: http://www-bcf.usc.edu/~gareth/ISL/" width="90%" />
<p class="caption">
Figure 6.4: Left: two classes of observations and a maximum margin hyperplane (solid line). Right: Hyperplane (solid line) moves after the addition of a new observation (original hyperplane is dashed line). Source: <a href="http://www-bcf.usc.edu/~gareth/ISL/" class="uri">http://www-bcf.usc.edu/~gareth/ISL/</a>
</p>
</div>
<p>It would be better to choose a classifier based on a hyperplane that:</p>
<ul>
<li>is more robust to individual observations</li>
<li>provides better classification of most of the training variables</li>
</ul>
<p>In other words, we might tolerate some misclassifications if the prediction of the remaining observations is more reliable. The <strong>support vector classifier</strong> does this by allowing some observations to be on the wrong side of the margin or even on the wrong side of the hyperplane. Observations on the wrong side of the hyperplane are misclassifications.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmObsOnWrongSideHyperplane"></span>
<img src="images/svm.9.6.png" alt="Left: observations on the wrong side of the margin. Right: observations on the wrong side of the margin and observations on the wrong side of the hyperplane. Source: http://www-bcf.usc.edu/~gareth/ISL/" width="90%" />
<p class="caption">
Figure 6.5: Left: observations on the wrong side of the margin. Right: observations on the wrong side of the margin and observations on the wrong side of the hyperplane. Source: <a href="http://www-bcf.usc.edu/~gareth/ISL/" class="uri">http://www-bcf.usc.edu/~gareth/ISL/</a>
</p>
</div>
<p>The support vector classifier has a tuning parameter, <em>C</em>, that determines the number and severity of the violations to the margin. If <em>C</em> = 0, then no violations to the margin will be tolerated, which is equivalent to the maximal margin classifier. As <em>C</em> increases, the classifier becomes more tolerant of violations to the margin, and so the margin widens.</p>
<p>The optimal value of <em>C</em> is chosen through cross-validation.</p>
<p><em>C</em> is described as a tuning parameter, because it controls the bias-variance trade-off:</p>
<ul>
<li>a small <em>C</em> results in narrow margins that are rarely violated; the model will have low bias, but high variance.</li>
<li>as <em>C</em> increases the margins widen allowing more violations; the bias of the model will increase, but its variance will decrease.</li>
</ul>
<p>The <strong>support vectors</strong> are the observations that lie directly on the margin, or on the wrong side of the margin for their class. The only observations that affect the classifier are the support vectors. As <em>C</em> increases, the margin widens and the number of support vectors increases. In other words, when <em>C</em> increases more observations are involved in determining the decision boundary of the classifier.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmMarginC"></span>
<img src="images/svm.9.7.png" alt="Margin of a support vector classifier changing with tuning parameter C. Largest value of C was used in the top left panel, and smaller values in the top right, bottom left and bottom right panels. Source: http://www-bcf.usc.edu/~gareth/ISL/" width="75%" />
<p class="caption">
Figure 6.6: Margin of a support vector classifier changing with tuning parameter C. Largest value of C was used in the top left panel, and smaller values in the top right, bottom left and bottom right panels. Source: <a href="http://www-bcf.usc.edu/~gareth/ISL/" class="uri">http://www-bcf.usc.edu/~gareth/ISL/</a>
</p>
</div>
</div>
<div id="support-vector-machine" class="section level2 hasAnchor" number="6.3">
<h2><span class="header-section-number">6.3</span> Support Vector Machine<a href="svm.html#support-vector-machine" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The support vector classifier performs well if we have linearly separable classes, however this isn’t always the case.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmNonLinearBoundary"></span>
<img src="images/svm.9.8.png" alt="Two classes of observations with a non-linear boundary between them." width="90%" />
<p class="caption">
Figure 6.7: Two classes of observations with a non-linear boundary between them.
</p>
</div>
<p>The SVM uses the <strong>kernel trick</strong> to operate in a higher dimensional space, without ever computing the coordinates of the data in that space.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmKernelMachine"></span>
<img src="images/svm_kernel_machine.png" alt="Kernel machine. By Alisneaky - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=14941564" width="80%" />
<p class="caption">
Figure 6.8: Kernel machine. By Alisneaky - Own work, CC0, <a href="https://commons.wikimedia.org/w/index.php?curid=14941564" class="uri">https://commons.wikimedia.org/w/index.php?curid=14941564</a>
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmPolyAndRadialKernelSVM"></span>
<img src="images/svm.9.9.png" alt="Left: SVM with polynomial kernel of degree 3. Right: SVM with radial kernel. Source: http://www-bcf.usc.edu/~gareth/ISL/" width="90%" />
<p class="caption">
Figure 6.9: Left: SVM with polynomial kernel of degree 3. Right: SVM with radial kernel. Source: <a href="http://www-bcf.usc.edu/~gareth/ISL/" class="uri">http://www-bcf.usc.edu/~gareth/ISL/</a>
</p>
</div>
</div>
<div id="example---training-a-classifier" class="section level2 hasAnchor" number="6.4">
<h2><span class="header-section-number">6.4</span> Example - training a classifier<a href="svm.html#example---training-a-classifier" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Training of an SVM will be demonstrated on a 2-dimensional simulated data set, with a non-linear decision boundary.</p>
<div id="setup-environment" class="section level3 hasAnchor" number="6.4.1">
<h3><span class="header-section-number">6.4.1</span> Setup environment<a href="svm.html#setup-environment" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Load required libraries</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="svm.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caret)</span></code></pre></div>
<pre><code>## Warning in system(&quot;timedatectl&quot;, intern = TRUE): running command &#39;timedatectl&#39;
## had status 1</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="svm.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb3-2"><a href="svm.html#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb3-3"><a href="svm.html#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pROC)</span>
<span id="cb3-4"><a href="svm.html#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(e1071)</span></code></pre></div>
</div>
<div id="partition-data-1" class="section level3 hasAnchor" number="6.4.2">
<h3><span class="header-section-number">6.4.2</span> Partition data<a href="svm.html#partition-data-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Load data</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="svm.html#cb4-1" aria-hidden="true" tabindex="-1"></a>moons <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/sim_data_svm/moons.csv&quot;</span>, <span class="at">header=</span>F)</span>
<span id="cb4-2"><a href="svm.html#cb4-2" aria-hidden="true" tabindex="-1"></a>moons<span class="sc">$</span>V3<span class="ot">=</span><span class="fu">as.factor</span>(moons<span class="sc">$</span>V3)</span>
<span id="cb4-3"><a href="svm.html#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(moons)</span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    400 obs. of  3 variables:
##  $ V1: num  -0.496 1.827 1.322 -1.138 -0.21 ...
##  $ V2: num  0.985 -0.501 -0.397 0.192 -0.145 ...
##  $ V3: Factor w/ 2 levels &quot;A&quot;,&quot;B&quot;: 1 2 2 1 2 1 1 2 1 2 ...</code></pre>
<p>V1 and V2 are the predictors; V3 is the class.</p>
<p>Partition data into training and test set</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="svm.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb6-2"><a href="svm.html#cb6-2" aria-hidden="true" tabindex="-1"></a>trainIndex <span class="ot">&lt;-</span> <span class="fu">createDataPartition</span>(<span class="at">y=</span>moons<span class="sc">$</span>V3, <span class="at">times=</span><span class="dv">1</span>, <span class="at">p=</span><span class="fl">0.7</span>, <span class="at">list=</span>F)</span>
<span id="cb6-3"><a href="svm.html#cb6-3" aria-hidden="true" tabindex="-1"></a>moonsTrain <span class="ot">&lt;-</span> moons[trainIndex,]</span>
<span id="cb6-4"><a href="svm.html#cb6-4" aria-hidden="true" tabindex="-1"></a>moonsTest <span class="ot">&lt;-</span> moons[<span class="sc">-</span>trainIndex,]</span>
<span id="cb6-5"><a href="svm.html#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="svm.html#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(moonsTrain<span class="sc">$</span>V3)</span></code></pre></div>
<pre><code>##   A   B 
## 140 140</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="svm.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(moonsTest<span class="sc">$</span>V3)</span></code></pre></div>
<pre><code>##  A  B 
## 60 60</code></pre>
</div>
<div id="visualize-training-data" class="section level3 hasAnchor" number="6.4.3">
<h3><span class="header-section-number">6.4.3</span> Visualize training data<a href="svm.html#visualize-training-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="svm.html#cb10-1" aria-hidden="true" tabindex="-1"></a>point_shapes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15</span>,<span class="dv">17</span>)</span>
<span id="cb10-2"><a href="svm.html#cb10-2" aria-hidden="true" tabindex="-1"></a>bp <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">3</span>,<span class="st">&quot;Dark2&quot;</span>)</span>
<span id="cb10-3"><a href="svm.html#cb10-3" aria-hidden="true" tabindex="-1"></a>point_colours <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(moonsTrain<span class="sc">$</span>V3<span class="sc">==</span><span class="st">&quot;A&quot;</span>, bp[<span class="dv">1</span>], bp[<span class="dv">2</span>])</span>
<span id="cb10-4"><a href="svm.html#cb10-4" aria-hidden="true" tabindex="-1"></a>point_shapes <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(moonsTrain<span class="sc">$</span>V3<span class="sc">==</span><span class="st">&quot;A&quot;</span>, <span class="dv">15</span>, <span class="dv">17</span>)</span>
<span id="cb10-5"><a href="svm.html#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="svm.html#cb10-6" aria-hidden="true" tabindex="-1"></a>point_size <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb10-7"><a href="svm.html#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="svm.html#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(moonsTrain, <span class="fu">aes</span>(V1,V2)) <span class="sc">+</span> </span>
<span id="cb10-9"><a href="svm.html#cb10-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">col=</span>point_colours, <span class="at">shape=</span>point_shapes, </span>
<span id="cb10-10"><a href="svm.html#cb10-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">size=</span>point_size) <span class="sc">+</span> </span>
<span id="cb10-11"><a href="svm.html#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb10-12"><a href="svm.html#cb10-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">25</span>, <span class="at">face=</span><span class="st">&quot;bold&quot;</span>), <span class="at">axis.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">15</span>),</span>
<span id="cb10-13"><a href="svm.html#cb10-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">20</span>,<span class="at">face=</span><span class="st">&quot;bold&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmMoonsTrainSet"></span>
<img src="05-support-vector-machines_files/figure-html/svmMoonsTrainSet-1.png" alt="Scatterplot of the training data" width="50%" />
<p class="caption">
Figure 6.10: Scatterplot of the training data
</p>
</div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="svm.html#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb11-2"><a href="svm.html#cb11-2" aria-hidden="true" tabindex="-1"></a>seeds <span class="ot">=</span> <span class="fu">vector</span>(<span class="at">mode=</span><span class="st">&#39;list&#39;</span>,<span class="at">length=</span><span class="dv">101</span>) <span class="co">#this is #folds+1 so 10+1</span></span>
<span id="cb11-3"><a href="svm.html#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>) seeds[[i]] <span class="ot">=</span> <span class="fu">sample.int</span>(<span class="dv">1000</span>,<span class="dv">100</span>)</span>
<span id="cb11-4"><a href="svm.html#cb11-4" aria-hidden="true" tabindex="-1"></a>seeds[[<span class="dv">101</span>]] <span class="ot">=</span> <span class="fu">sample.int</span>(<span class="dv">1000</span>,<span class="dv">1</span>)</span>
<span id="cb11-5"><a href="svm.html#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="svm.html#cb11-6" aria-hidden="true" tabindex="-1"></a>trctrl <span class="ot">&lt;-</span> <span class="fu">trainControl</span>(<span class="at">method =</span> <span class="st">&quot;repeatedcv&quot;</span>, </span>
<span id="cb11-7"><a href="svm.html#cb11-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">number =</span> <span class="dv">10</span>, </span>
<span id="cb11-8"><a href="svm.html#cb11-8" aria-hidden="true" tabindex="-1"></a>                       <span class="at">repeats =</span> <span class="dv">10</span>,</span>
<span id="cb11-9"><a href="svm.html#cb11-9" aria-hidden="true" tabindex="-1"></a>                       <span class="at">seeds =</span> seeds,</span>
<span id="cb11-10"><a href="svm.html#cb11-10" aria-hidden="true" tabindex="-1"></a>                       <span class="at">classProbs=</span><span class="cn">TRUE</span>)</span>
<span id="cb11-11"><a href="svm.html#cb11-11" aria-hidden="true" tabindex="-1"></a>svmTune <span class="ot">&lt;-</span> <span class="fu">train</span>(<span class="at">x =</span> moonsTrain[,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)],</span>
<span id="cb11-12"><a href="svm.html#cb11-12" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y =</span> moonsTrain[,<span class="dv">3</span>],</span>
<span id="cb11-13"><a href="svm.html#cb11-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">method =</span> <span class="st">&quot;svmRadial&quot;</span>,</span>
<span id="cb11-14"><a href="svm.html#cb11-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">tuneLength =</span> <span class="dv">9</span>,</span>
<span id="cb11-15"><a href="svm.html#cb11-15" aria-hidden="true" tabindex="-1"></a>                 <span class="at">preProc =</span> <span class="fu">c</span>(<span class="st">&quot;center&quot;</span>, <span class="st">&quot;scale&quot;</span>),</span>
<span id="cb11-16"><a href="svm.html#cb11-16" aria-hidden="true" tabindex="-1"></a>                 <span class="at">trControl =</span> trctrl)</span>
<span id="cb11-17"><a href="svm.html#cb11-17" aria-hidden="true" tabindex="-1"></a>                 </span>
<span id="cb11-18"><a href="svm.html#cb11-18" aria-hidden="true" tabindex="-1"></a>svmTune</span></code></pre></div>
<pre><code>## Support Vector Machines with Radial Basis Function Kernel 
## 
## 280 samples
##   2 predictor
##   2 classes: &#39;A&#39;, &#39;B&#39; 
## 
## Pre-processing: centered (2), scaled (2) 
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 252, 252, 252, 252, 252, 252, ... 
## Resampling results across tuning parameters:
## 
##   C      Accuracy   Kappa    
##    0.25  0.8875000  0.7750000
##    0.50  0.8846429  0.7692857
##    1.00  0.8810714  0.7621429
##    2.00  0.8803571  0.7607143
##    4.00  0.8750000  0.7500000
##    8.00  0.8671429  0.7342857
##   16.00  0.8635714  0.7271429
##   32.00  0.8582143  0.7164286
##   64.00  0.8557143  0.7114286
## 
## Tuning parameter &#39;sigma&#39; was held constant at a value of 1.36437
## Accuracy was used to select the optimal model using the largest value.
## The final values used for the model were sigma = 1.36437 and C = 0.25.</code></pre>
</div>
<div id="prediction-performance-measures" class="section level3 hasAnchor" number="6.4.4">
<h3><span class="header-section-number">6.4.4</span> Prediction performance measures<a href="svm.html#prediction-performance-measures" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>SVM accuracy profile</p>
<p>Predictions on test set.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="svm.html#cb13-1" aria-hidden="true" tabindex="-1"></a>svmPred <span class="ot">&lt;-</span> <span class="fu">predict</span>(svmTune, moonsTest[,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)])</span>
<span id="cb13-2"><a href="svm.html#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">confusionMatrix</span>(svmPred, <span class="fu">as.factor</span>(moonsTest[,<span class="dv">3</span>]))</span></code></pre></div>
<pre><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction  A  B
##          A 59  7
##          B  1 53
##                                           
##                Accuracy : 0.9333          
##                  95% CI : (0.8729, 0.9708)
##     No Information Rate : 0.5             
##     P-Value [Acc &gt; NIR] : &lt;2e-16          
##                                           
##                   Kappa : 0.8667          
##                                           
##  Mcnemar&#39;s Test P-Value : 0.0771          
##                                           
##             Sensitivity : 0.9833          
##             Specificity : 0.8833          
##          Pos Pred Value : 0.8939          
##          Neg Pred Value : 0.9815          
##              Prevalence : 0.5000          
##          Detection Rate : 0.4917          
##    Detection Prevalence : 0.5500          
##       Balanced Accuracy : 0.9333          
##                                           
##        &#39;Positive&#39; Class : A               
## </code></pre>
<p>Get predicted class probabilities so we can build ROC curve.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="svm.html#cb15-1" aria-hidden="true" tabindex="-1"></a>svmProbs <span class="ot">&lt;-</span> <span class="fu">predict</span>(svmTune, moonsTest[,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)], <span class="at">type=</span><span class="st">&quot;prob&quot;</span>)</span>
<span id="cb15-2"><a href="svm.html#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(svmProbs)</span></code></pre></div>
<pre><code>##            A          B
## 1 0.06967502 0.93032498
## 2 0.93443571 0.06556429
## 3 0.93707422 0.06292578
## 4 0.07198768 0.92801232
## 5 0.88761226 0.11238774
## 6 0.94212490 0.05787510</code></pre>
<p>Build a ROC curve.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="svm.html#cb17-1" aria-hidden="true" tabindex="-1"></a>svmROC <span class="ot">&lt;-</span> <span class="fu">roc</span>(moonsTest[,<span class="dv">3</span>], svmProbs[,<span class="st">&quot;A&quot;</span>])</span></code></pre></div>
<pre><code>## Setting levels: control = A, case = B</code></pre>
<pre><code>## Setting direction: controls &gt; cases</code></pre>
<p>Plot ROC curve.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="svm.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(svmROC, <span class="at">type =</span> <span class="st">&quot;S&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmROCcurveMoons"></span>
<img src="05-support-vector-machines_files/figure-html/svmROCcurveMoons-1.png" alt="SVM accuracy profile." width="80%" />
<p class="caption">
Figure 6.11: SVM accuracy profile.
</p>
</div>
<p><strong>Sensitivity (true positive rate)</strong></p>
<p><em>TPR = TP/P = TP/(TP+FN)</em></p>
<p><strong>Specificity (true negative rate)</strong></p>
<p><em>SPC = TN/N = TN/(TN+FP)</em></p>
<p>Calculate area under ROC curve.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="svm.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">auc</span>(svmROC)</span></code></pre></div>
<pre><code>## Area under the curve: 0.9728</code></pre>
</div>
<div id="plot-decision-boundary" class="section level3 hasAnchor" number="6.4.5">
<h3><span class="header-section-number">6.4.5</span> Plot decision boundary<a href="svm.html#plot-decision-boundary" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Create a grid so we can predict across the full range of our variables V1 and V2.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="svm.html#cb23-1" aria-hidden="true" tabindex="-1"></a>gridSize <span class="ot">&lt;-</span> <span class="dv">150</span> </span>
<span id="cb23-2"><a href="svm.html#cb23-2" aria-hidden="true" tabindex="-1"></a>v1limits <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">min</span>(moons<span class="sc">$</span>V1),<span class="fu">max</span>(moons<span class="sc">$</span>V1))</span>
<span id="cb23-3"><a href="svm.html#cb23-3" aria-hidden="true" tabindex="-1"></a>tmpV1 <span class="ot">&lt;-</span> <span class="fu">seq</span>(v1limits[<span class="dv">1</span>],v1limits[<span class="dv">2</span>],<span class="at">len=</span>gridSize)</span>
<span id="cb23-4"><a href="svm.html#cb23-4" aria-hidden="true" tabindex="-1"></a>v2limits <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">min</span>(moons<span class="sc">$</span>V2), <span class="fu">max</span>(moons<span class="sc">$</span>V2))</span>
<span id="cb23-5"><a href="svm.html#cb23-5" aria-hidden="true" tabindex="-1"></a>tmpV2 <span class="ot">&lt;-</span> <span class="fu">seq</span>(v2limits[<span class="dv">1</span>],v2limits[<span class="dv">2</span>],<span class="at">len=</span>gridSize)</span>
<span id="cb23-6"><a href="svm.html#cb23-6" aria-hidden="true" tabindex="-1"></a>xgrid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(tmpV1,tmpV2)</span>
<span id="cb23-7"><a href="svm.html#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(xgrid) <span class="ot">&lt;-</span> <span class="fu">names</span>(moons)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code></pre></div>
<p>Predict values of all elements of grid.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="svm.html#cb24-1" aria-hidden="true" tabindex="-1"></a>V3 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">predict</span>(svmTune, xgrid))</span>
<span id="cb24-2"><a href="svm.html#cb24-2" aria-hidden="true" tabindex="-1"></a>xgrid <span class="ot">&lt;-</span> <span class="fu">cbind</span>(xgrid, V3)</span></code></pre></div>
<p>Plot</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="svm.html#cb25-1" aria-hidden="true" tabindex="-1"></a>point_shapes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15</span>,<span class="dv">17</span>)</span>
<span id="cb25-2"><a href="svm.html#cb25-2" aria-hidden="true" tabindex="-1"></a>point_colours <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">3</span>,<span class="st">&quot;Dark2&quot;</span>)</span>
<span id="cb25-3"><a href="svm.html#cb25-3" aria-hidden="true" tabindex="-1"></a>point_size <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb25-4"><a href="svm.html#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="svm.html#cb25-5" aria-hidden="true" tabindex="-1"></a>trainClassNumeric <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(moonsTrain<span class="sc">$</span>V3<span class="sc">==</span><span class="st">&quot;A&quot;</span>, <span class="dv">1</span>, <span class="dv">2</span>)</span>
<span id="cb25-6"><a href="svm.html#cb25-6" aria-hidden="true" tabindex="-1"></a>testClassNumeric <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(moonsTest<span class="sc">$</span>V3<span class="sc">==</span><span class="st">&quot;A&quot;</span>, <span class="dv">1</span>, <span class="dv">2</span>)</span>
<span id="cb25-7"><a href="svm.html#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="svm.html#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(xgrid, <span class="fu">aes</span>(V1,V2)) <span class="sc">+</span></span>
<span id="cb25-9"><a href="svm.html#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">col=</span>point_colours[V3], <span class="at">shape=</span><span class="dv">16</span>, <span class="at">size=</span><span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb25-10"><a href="svm.html#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>moonsTrain, <span class="fu">aes</span>(V1,V2), <span class="at">col=</span>point_colours[trainClassNumeric],</span>
<span id="cb25-11"><a href="svm.html#cb25-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">shape=</span>point_shapes[trainClassNumeric], <span class="at">size=</span>point_size) <span class="sc">+</span></span>
<span id="cb25-12"><a href="svm.html#cb25-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_contour</span>(<span class="at">data=</span>xgrid, <span class="fu">aes</span>(<span class="at">x=</span>V1, <span class="at">y=</span>V2, <span class="at">z=</span>V3), <span class="at">breaks=</span><span class="fl">1.5</span>, <span class="at">col=</span><span class="st">&quot;grey30&quot;</span>) <span class="sc">+</span></span>
<span id="cb25-13"><a href="svm.html#cb25-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;train&quot;</span>) <span class="sc">+</span></span>
<span id="cb25-14"><a href="svm.html#cb25-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb25-15"><a href="svm.html#cb25-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">25</span>, <span class="at">face=</span><span class="st">&quot;bold&quot;</span>), <span class="at">axis.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">15</span>),</span>
<span id="cb25-16"><a href="svm.html#cb25-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">20</span>,<span class="at">face=</span><span class="st">&quot;bold&quot;</span>))</span>
<span id="cb25-17"><a href="svm.html#cb25-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-18"><a href="svm.html#cb25-18" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(xgrid, <span class="fu">aes</span>(V1,V2)) <span class="sc">+</span></span>
<span id="cb25-19"><a href="svm.html#cb25-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">col=</span>point_colours[V3], <span class="at">shape=</span><span class="dv">16</span>, <span class="at">size=</span><span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb25-20"><a href="svm.html#cb25-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>moonsTest, <span class="fu">aes</span>(V1,V2), <span class="at">col=</span>point_colours[testClassNumeric],</span>
<span id="cb25-21"><a href="svm.html#cb25-21" aria-hidden="true" tabindex="-1"></a>             <span class="at">shape=</span>point_shapes[testClassNumeric], <span class="at">size=</span>point_size) <span class="sc">+</span></span>
<span id="cb25-22"><a href="svm.html#cb25-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_contour</span>(<span class="at">data=</span>xgrid, <span class="fu">aes</span>(<span class="at">x=</span>V1, <span class="at">y=</span>V2, <span class="at">z=</span>V3), <span class="at">breaks=</span><span class="fl">1.5</span>, <span class="at">col=</span><span class="st">&quot;grey30&quot;</span>) <span class="sc">+</span></span>
<span id="cb25-23"><a href="svm.html#cb25-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;test&quot;</span>) <span class="sc">+</span></span>
<span id="cb25-24"><a href="svm.html#cb25-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb25-25"><a href="svm.html#cb25-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">25</span>, <span class="at">face=</span><span class="st">&quot;bold&quot;</span>), <span class="at">axis.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">15</span>),</span>
<span id="cb25-26"><a href="svm.html#cb25-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">20</span>,<span class="at">face=</span><span class="st">&quot;bold&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:simDataBinClassDecisionBoundarySVM"></span>
<img src="05-support-vector-machines_files/figure-html/simDataBinClassDecisionBoundarySVM-1.png" alt="Decision boundary created by radial kernel SVM." width="50%" /><img src="05-support-vector-machines_files/figure-html/simDataBinClassDecisionBoundarySVM-2.png" alt="Decision boundary created by radial kernel SVM." width="50%" />
<p class="caption">
Figure 6.12: Decision boundary created by radial kernel SVM.
</p>
</div>
</div>
</div>
<div id="defining-your-own-model-type-to-use-in-caret" class="section level2 hasAnchor" number="6.5">
<h2><span class="header-section-number">6.5</span> Defining your own model type to use in caret<a href="svm.html#defining-your-own-model-type-to-use-in-caret" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Caret has over two hundred built in models, including several support vector machines:
<a href="https://topepo.github.io/caret/available-models.html">https://topepo.github.io/caret/available-models.html</a></p>
<p>However, despite this wide range of options, you may occasionally need to define your own model. Caret does not currently have a radial SVM implemented using the <a href="https://cran.r-project.org/package=e1071">e1071 library</a>, so we will define one here.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="svm.html#cb26-1" aria-hidden="true" tabindex="-1"></a>svmRadialE1071 <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb26-2"><a href="svm.html#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> <span class="st">&quot;Support Vector Machines with Radial Kernel - e1071&quot;</span>,</span>
<span id="cb26-3"><a href="svm.html#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">library =</span> <span class="st">&quot;e1071&quot;</span>,</span>
<span id="cb26-4"><a href="svm.html#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="fu">c</span>(<span class="st">&quot;Regression&quot;</span>, <span class="st">&quot;Classification&quot;</span>),</span>
<span id="cb26-5"><a href="svm.html#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">parameters =</span> <span class="fu">data.frame</span>(<span class="at">parameter=</span><span class="st">&quot;cost&quot;</span>,</span>
<span id="cb26-6"><a href="svm.html#cb26-6" aria-hidden="true" tabindex="-1"></a>                          <span class="at">class=</span><span class="st">&quot;numeric&quot;</span>,</span>
<span id="cb26-7"><a href="svm.html#cb26-7" aria-hidden="true" tabindex="-1"></a>                          <span class="at">label=</span><span class="st">&quot;Cost&quot;</span>),</span>
<span id="cb26-8"><a href="svm.html#cb26-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">grid =</span> <span class="cf">function</span> (x, y, <span class="at">len =</span> <span class="cn">NULL</span>, <span class="at">search =</span> <span class="st">&quot;grid&quot;</span>) </span>
<span id="cb26-9"><a href="svm.html#cb26-9" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb26-10"><a href="svm.html#cb26-10" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (search <span class="sc">==</span> <span class="st">&quot;grid&quot;</span>) {</span>
<span id="cb26-11"><a href="svm.html#cb26-11" aria-hidden="true" tabindex="-1"></a>        out <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">cost =</span> <span class="dv">2</span><span class="sc">^</span>((<span class="dv">1</span><span class="sc">:</span>len) <span class="sc">-</span> <span class="dv">3</span>))</span>
<span id="cb26-12"><a href="svm.html#cb26-12" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb26-13"><a href="svm.html#cb26-13" aria-hidden="true" tabindex="-1"></a>      <span class="cf">else</span> {</span>
<span id="cb26-14"><a href="svm.html#cb26-14" aria-hidden="true" tabindex="-1"></a>        out <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">cost =</span> <span class="dv">2</span><span class="sc">^</span><span class="fu">runif</span>(len, <span class="at">min =</span> <span class="sc">-</span><span class="dv">5</span>, <span class="at">max =</span> <span class="dv">10</span>))</span>
<span id="cb26-15"><a href="svm.html#cb26-15" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb26-16"><a href="svm.html#cb26-16" aria-hidden="true" tabindex="-1"></a>      out</span>
<span id="cb26-17"><a href="svm.html#cb26-17" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb26-18"><a href="svm.html#cb26-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">loop=</span><span class="cn">NULL</span>,</span>
<span id="cb26-19"><a href="svm.html#cb26-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">fit=</span><span class="cf">function</span> (x, y, wts, param, lev, last, classProbs, ...) </span>
<span id="cb26-20"><a href="svm.html#cb26-20" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb26-21"><a href="svm.html#cb26-21" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">any</span>(<span class="fu">names</span>(<span class="fu">list</span>(...)) <span class="sc">==</span> <span class="st">&quot;probability&quot;</span>) <span class="sc">|</span> <span class="fu">is.numeric</span>(y)) {</span>
<span id="cb26-22"><a href="svm.html#cb26-22" aria-hidden="true" tabindex="-1"></a>        out <span class="ot">&lt;-</span> e1071<span class="sc">::</span><span class="fu">svm</span>(<span class="at">x =</span> <span class="fu">as.matrix</span>(x), <span class="at">y =</span> y, <span class="at">kernel =</span> <span class="st">&quot;radial&quot;</span>, </span>
<span id="cb26-23"><a href="svm.html#cb26-23" aria-hidden="true" tabindex="-1"></a>                          <span class="at">cost =</span> param<span class="sc">$</span>cost, ...)</span>
<span id="cb26-24"><a href="svm.html#cb26-24" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb26-25"><a href="svm.html#cb26-25" aria-hidden="true" tabindex="-1"></a>      <span class="cf">else</span> {</span>
<span id="cb26-26"><a href="svm.html#cb26-26" aria-hidden="true" tabindex="-1"></a>        out <span class="ot">&lt;-</span> e1071<span class="sc">::</span><span class="fu">svm</span>(<span class="at">x =</span> <span class="fu">as.matrix</span>(x), <span class="at">y =</span> y, <span class="at">kernel =</span> <span class="st">&quot;radial&quot;</span>, </span>
<span id="cb26-27"><a href="svm.html#cb26-27" aria-hidden="true" tabindex="-1"></a>                          <span class="at">cost =</span> param<span class="sc">$</span>cost, <span class="at">probability =</span> classProbs, ...)</span>
<span id="cb26-28"><a href="svm.html#cb26-28" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb26-29"><a href="svm.html#cb26-29" aria-hidden="true" tabindex="-1"></a>      out</span>
<span id="cb26-30"><a href="svm.html#cb26-30" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb26-31"><a href="svm.html#cb26-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">predict =</span> <span class="cf">function</span> (modelFit, newdata, <span class="at">submodels =</span> <span class="cn">NULL</span>) </span>
<span id="cb26-32"><a href="svm.html#cb26-32" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb26-33"><a href="svm.html#cb26-33" aria-hidden="true" tabindex="-1"></a>      <span class="fu">predict</span>(modelFit, newdata)</span>
<span id="cb26-34"><a href="svm.html#cb26-34" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb26-35"><a href="svm.html#cb26-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">prob =</span> <span class="cf">function</span> (modelFit, newdata, <span class="at">submodels =</span> <span class="cn">NULL</span>) </span>
<span id="cb26-36"><a href="svm.html#cb26-36" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb26-37"><a href="svm.html#cb26-37" aria-hidden="true" tabindex="-1"></a>      out <span class="ot">&lt;-</span> <span class="fu">predict</span>(modelFit, newdata, <span class="at">probability =</span> <span class="cn">TRUE</span>)</span>
<span id="cb26-38"><a href="svm.html#cb26-38" aria-hidden="true" tabindex="-1"></a>      <span class="fu">attr</span>(out, <span class="st">&quot;probabilities&quot;</span>)</span>
<span id="cb26-39"><a href="svm.html#cb26-39" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb26-40"><a href="svm.html#cb26-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">predictors =</span> <span class="cf">function</span> (x, ...) </span>
<span id="cb26-41"><a href="svm.html#cb26-41" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb26-42"><a href="svm.html#cb26-42" aria-hidden="true" tabindex="-1"></a>      out <span class="ot">&lt;-</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(x<span class="sc">$</span>terms)) </span>
<span id="cb26-43"><a href="svm.html#cb26-43" aria-hidden="true" tabindex="-1"></a>        <span class="fu">predictors.terms</span>(x<span class="sc">$</span>terms)</span>
<span id="cb26-44"><a href="svm.html#cb26-44" aria-hidden="true" tabindex="-1"></a>      <span class="cf">else</span> x<span class="sc">$</span>xNames</span>
<span id="cb26-45"><a href="svm.html#cb26-45" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">is.null</span>(out)) </span>
<span id="cb26-46"><a href="svm.html#cb26-46" aria-hidden="true" tabindex="-1"></a>        out <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">attr</span>(x, <span class="st">&quot;scaling&quot;</span>)<span class="sc">$</span>x.scale<span class="sc">$</span><span class="st">`</span><span class="at">scaled:center</span><span class="st">`</span>)</span>
<span id="cb26-47"><a href="svm.html#cb26-47" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">is.null</span>(out)) </span>
<span id="cb26-48"><a href="svm.html#cb26-48" aria-hidden="true" tabindex="-1"></a>        out <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb26-49"><a href="svm.html#cb26-49" aria-hidden="true" tabindex="-1"></a>      out</span>
<span id="cb26-50"><a href="svm.html#cb26-50" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb26-51"><a href="svm.html#cb26-51" aria-hidden="true" tabindex="-1"></a>  <span class="at">tags =</span> <span class="fu">c</span>(<span class="st">&quot;Kernel Methods&quot;</span>, <span class="st">&quot;Support Vector Machines&quot;</span>, <span class="st">&quot;Regression&quot;</span>, <span class="st">&quot;Classifier&quot;</span>, <span class="st">&quot;Robust Methods&quot;</span>),</span>
<span id="cb26-52"><a href="svm.html#cb26-52" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="cf">function</span>(x) x<span class="sc">$</span>levels,</span>
<span id="cb26-53"><a href="svm.html#cb26-53" aria-hidden="true" tabindex="-1"></a>  <span class="at">sort =</span> <span class="cf">function</span>(x)</span>
<span id="cb26-54"><a href="svm.html#cb26-54" aria-hidden="true" tabindex="-1"></a>  {</span>
<span id="cb26-55"><a href="svm.html#cb26-55" aria-hidden="true" tabindex="-1"></a>    x[<span class="fu">order</span>(x<span class="sc">$</span>cost), ]</span>
<span id="cb26-56"><a href="svm.html#cb26-56" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb26-57"><a href="svm.html#cb26-57" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Note that the radial SVM model we have defined has only one tuning parameter, cost (<em>C</em>). If we do not define the kernel parameter <em>gamma</em>, e1071 will automatically calculate it as 1/(data dimension); <em>i.e.</em> if we have 58 predictors, <em>gamma</em> will be 1/58 = 0.01724.</p>
<div id="model-cross-validation-and-tuning" class="section level3 hasAnchor" number="6.5.1">
<h3><span class="header-section-number">6.5.1</span> Model cross-validation and tuning<a href="svm.html#model-cross-validation-and-tuning" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will pass the twoClassSummary function into model training through trainControl. Additionally we would like the model to predict class probabilities so that we can calculate the ROC curve, so we use the classProbs option.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="svm.html#cb27-1" aria-hidden="true" tabindex="-1"></a>cvCtrl <span class="ot">&lt;-</span> <span class="fu">trainControl</span>(<span class="at">method =</span> <span class="st">&quot;repeatedcv&quot;</span>, </span>
<span id="cb27-2"><a href="svm.html#cb27-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">repeats =</span> <span class="dv">10</span>,</span>
<span id="cb27-3"><a href="svm.html#cb27-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">number =</span> <span class="dv">10</span>,</span>
<span id="cb27-4"><a href="svm.html#cb27-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">summaryFunction =</span> twoClassSummary,</span>
<span id="cb27-5"><a href="svm.html#cb27-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">classProbs =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-6"><a href="svm.html#cb27-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">seeds=</span>seeds)</span></code></pre></div>
<p>We set the <strong>method</strong> of the <strong>train</strong> function to <strong>svmRadial</strong> to specify a radial kernel SVM. In this implementation we only have to tune one parameter, <strong>cost</strong>. The default grid of cost parameters start at 0.25 and double at each iteration. Choosing tuneLength = 9 will give us cost parameters of 0.25, 0.5, 1, 2, 4, 8, 16, 32 and 64.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="svm.html#cb28-1" aria-hidden="true" tabindex="-1"></a>svmTune <span class="ot">&lt;-</span> <span class="fu">train</span>(<span class="at">x =</span> moonsTrain[,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)],</span>
<span id="cb28-2"><a href="svm.html#cb28-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y =</span> moonsTrain[,<span class="dv">3</span>],</span>
<span id="cb28-3"><a href="svm.html#cb28-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">method =</span> svmRadialE1071,</span>
<span id="cb28-4"><a href="svm.html#cb28-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">tuneLength =</span> <span class="dv">9</span>,</span>
<span id="cb28-5"><a href="svm.html#cb28-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">preProc =</span> <span class="fu">c</span>(<span class="st">&quot;center&quot;</span>, <span class="st">&quot;scale&quot;</span>),</span>
<span id="cb28-6"><a href="svm.html#cb28-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">metric =</span> <span class="st">&quot;ROC&quot;</span>,</span>
<span id="cb28-7"><a href="svm.html#cb28-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">trControl =</span> cvCtrl)</span>
<span id="cb28-8"><a href="svm.html#cb28-8" aria-hidden="true" tabindex="-1"></a>                 </span>
<span id="cb28-9"><a href="svm.html#cb28-9" aria-hidden="true" tabindex="-1"></a>svmTune</span></code></pre></div>
<pre><code>## Support Vector Machines with Radial Kernel - e1071 
## 
## 280 samples
##   2 predictor
##   2 classes: &#39;A&#39;, &#39;B&#39; 
## 
## Pre-processing: centered (2), scaled (2) 
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 252, 252, 252, 252, 252, 252, ... 
## Resampling results across tuning parameters:
## 
##   cost   ROC        Sens       Spec     
##    0.25  0.9230612  0.8378571  0.9057143
##    0.50  0.9320918  0.8592857  0.9057143
##    1.00  0.9418878  0.8692857  0.9000000
##    2.00  0.9455102  0.8792857  0.8964286
##    4.00  0.9479082  0.8785714  0.8914286
##    8.00  0.9485204  0.8678571  0.8857143
##   16.00  0.9492857  0.8757143  0.8800000
##   32.00  0.9490816  0.8764286  0.8728571
##   64.00  0.9463776  0.8778571  0.8707143
## 
## ROC was used to select the optimal model using the largest value.
## The final value used for the model was cost = 16.</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="svm.html#cb30-1" aria-hidden="true" tabindex="-1"></a>svmTune<span class="sc">$</span>finalModel</span></code></pre></div>
<pre><code>## 
## Call:
## svm.default(x = as.matrix(x), y = y, kernel = &quot;radial&quot;, cost = param$cost, 
##     probability = classProbs)
## 
## 
## Parameters:
##    SVM-Type:  C-classification 
##  SVM-Kernel:  radial 
##        cost:  16 
## 
## Number of Support Vectors:  79</code></pre>
</div>
</div>
<div id="iris-example" class="section level2 hasAnchor" number="6.6">
<h2><span class="header-section-number">6.6</span> Iris example<a href="svm.html#iris-example" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="svm.html#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(datasets)</span>
<span id="cb32-2"><a href="svm.html#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(iris) <span class="do">##loads the dataset, which can be accessed under the variable name iris</span></span>
<span id="cb32-3"><a href="svm.html#cb32-3" aria-hidden="true" tabindex="-1"></a>?iris <span class="do">##opens the documentation for the dataset</span></span>
<span id="cb32-4"><a href="svm.html#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(iris) <span class="do">##presents the 5 figure summary of the dataset</span></span></code></pre></div>
<pre><code>##   Sepal.Length    Sepal.Width     Petal.Length    Petal.Width   
##  Min.   :4.300   Min.   :2.000   Min.   :1.000   Min.   :0.100  
##  1st Qu.:5.100   1st Qu.:2.800   1st Qu.:1.600   1st Qu.:0.300  
##  Median :5.800   Median :3.000   Median :4.350   Median :1.300  
##  Mean   :5.843   Mean   :3.057   Mean   :3.758   Mean   :1.199  
##  3rd Qu.:6.400   3rd Qu.:3.300   3rd Qu.:5.100   3rd Qu.:1.800  
##  Max.   :7.900   Max.   :4.400   Max.   :6.900   Max.   :2.500  
##        Species  
##  setosa    :50  
##  versicolor:50  
##  virginica :50  
##                 
##                 
## </code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="svm.html#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(iris) <span class="do">##presents the structure of the iris dataframe</span></span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    150 obs. of  5 variables:
##  $ Sepal.Length: num  5.1 4.9 4.7 4.6 5 5.4 4.6 5 4.4 4.9 ...
##  $ Sepal.Width : num  3.5 3 3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 ...
##  $ Petal.Length: num  1.4 1.4 1.3 1.5 1.4 1.7 1.4 1.5 1.4 1.5 ...
##  $ Petal.Width : num  0.2 0.2 0.2 0.2 0.2 0.4 0.3 0.2 0.2 0.1 ...
##  $ Species     : Factor w/ 3 levels &quot;setosa&quot;,&quot;versicolor&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="svm.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb36-2"><a href="svm.html#cb36-2" aria-hidden="true" tabindex="-1"></a>trainTestPartition<span class="ot">&lt;-</span><span class="fu">createDataPartition</span>(<span class="at">y=</span>iris<span class="sc">$</span>Species, <span class="co">#the class label, caret ensures an even split of classes</span></span>
<span id="cb36-3"><a href="svm.html#cb36-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">p=</span><span class="fl">0.7</span>, <span class="co">#proportion of samples assigned to train</span></span>
<span id="cb36-4"><a href="svm.html#cb36-4" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">list=</span><span class="cn">FALSE</span>)</span>
<span id="cb36-5"><a href="svm.html#cb36-5" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(trainTestPartition)</span></code></pre></div>
<pre><code>##  int [1:105, 1] 1 2 3 4 5 7 8 10 11 12 ...
##  - attr(*, &quot;dimnames&quot;)=List of 2
##   ..$ : NULL
##   ..$ : chr &quot;Resample1&quot;</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="svm.html#cb38-1" aria-hidden="true" tabindex="-1"></a>iris.training <span class="ot">&lt;-</span> iris[ trainTestPartition,] <span class="co">#take the corresponding rows for training</span></span>
<span id="cb38-2"><a href="svm.html#cb38-2" aria-hidden="true" tabindex="-1"></a>iris.testing  <span class="ot">&lt;-</span> iris[<span class="sc">-</span>trainTestPartition,] <span class="co">#take the corresponding rows for testing by removing training rows</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="svm.html#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb39-2"><a href="svm.html#cb39-2" aria-hidden="true" tabindex="-1"></a>seeds <span class="ot">=</span> <span class="fu">vector</span>(<span class="at">mode=</span><span class="st">&#39;list&#39;</span>,<span class="at">length=</span><span class="dv">101</span>) <span class="co">#you need length #folds*#repeats + 1 so 10*10 + 1 here</span></span>
<span id="cb39-3"><a href="svm.html#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>) seeds[[i]] <span class="ot">=</span> <span class="fu">sample.int</span>(<span class="dv">1000</span>,<span class="dv">10</span>)</span>
<span id="cb39-4"><a href="svm.html#cb39-4" aria-hidden="true" tabindex="-1"></a>seeds[[<span class="dv">101</span>]] <span class="ot">=</span> <span class="fu">sample.int</span>(<span class="dv">1000</span>,<span class="dv">1</span>)</span>
<span id="cb39-5"><a href="svm.html#cb39-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-6"><a href="svm.html#cb39-6" aria-hidden="true" tabindex="-1"></a>train_ctrl_seed_repeated <span class="ot">=</span> <span class="fu">trainControl</span>(<span class="at">method=</span><span class="st">&#39;repeatedcv&#39;</span>,</span>
<span id="cb39-7"><a href="svm.html#cb39-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">number=</span><span class="dv">10</span>, <span class="co">#number of folds</span></span>
<span id="cb39-8"><a href="svm.html#cb39-8" aria-hidden="true" tabindex="-1"></a>                              <span class="at">repeats=</span><span class="dv">10</span>, <span class="co">#number of times to repeat cross-validation</span></span>
<span id="cb39-9"><a href="svm.html#cb39-9" aria-hidden="true" tabindex="-1"></a>                              <span class="at">seeds=</span>seeds)</span>
<span id="cb39-10"><a href="svm.html#cb39-10" aria-hidden="true" tabindex="-1"></a>iris_svm <span class="ot">&lt;-</span> <span class="fu">train</span>(</span>
<span id="cb39-11"><a href="svm.html#cb39-11" aria-hidden="true" tabindex="-1"></a>                  Species <span class="sc">~</span> .,</span>
<span id="cb39-12"><a href="svm.html#cb39-12" aria-hidden="true" tabindex="-1"></a>                  <span class="at">data =</span> iris.training,</span>
<span id="cb39-13"><a href="svm.html#cb39-13" aria-hidden="true" tabindex="-1"></a>                  <span class="at">method =</span> <span class="st">&quot;svmLinear&quot;</span>,</span>
<span id="cb39-14"><a href="svm.html#cb39-14" aria-hidden="true" tabindex="-1"></a>                  <span class="at">parms =</span> <span class="fu">list</span>(<span class="at">split =</span> <span class="st">&quot;gini&quot;</span>),</span>
<span id="cb39-15"><a href="svm.html#cb39-15" aria-hidden="true" tabindex="-1"></a>                  <span class="at">preProc =</span> <span class="fu">c</span>(<span class="st">&quot;corr&quot;</span>,<span class="st">&quot;nzv&quot;</span>,<span class="st">&quot;center&quot;</span>, <span class="st">&quot;scale&quot;</span>,<span class="st">&quot;BoxCox&quot;</span>),</span>
<span id="cb39-16"><a href="svm.html#cb39-16" aria-hidden="true" tabindex="-1"></a>                  <span class="at">tuneLength=</span><span class="dv">10</span>,</span>
<span id="cb39-17"><a href="svm.html#cb39-17" aria-hidden="true" tabindex="-1"></a>                  <span class="at">trControl =</span> train_ctrl_seed_repeated</span>
<span id="cb39-18"><a href="svm.html#cb39-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb39-19"><a href="svm.html#cb39-19" aria-hidden="true" tabindex="-1"></a>iris_svm</span></code></pre></div>
<pre><code>## Support Vector Machines with Linear Kernel 
## 
## 105 samples
##   4 predictor
##   3 classes: &#39;setosa&#39;, &#39;versicolor&#39;, &#39;virginica&#39; 
## 
## Pre-processing: centered (3), scaled (3), Box-Cox transformation (3), remove (1) 
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 94, 95, 94, 93, 95, 95, ... 
## Resampling results:
## 
##   Accuracy   Kappa    
##   0.9740152  0.9605099
## 
## Tuning parameter &#39;C&#39; was held constant at a value of 1</code></pre>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="svm.html#cb41-1" aria-hidden="true" tabindex="-1"></a>iris_information_predict_train<span class="ot">=</span><span class="fu">predict</span>(iris_svm,iris.training,<span class="at">type=</span><span class="st">&#39;raw&#39;</span>)</span>
<span id="cb41-2"><a href="svm.html#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">confusionMatrix</span>(iris_information_predict_train,iris.training<span class="sc">$</span>Species)</span></code></pre></div>
<pre><code>## Confusion Matrix and Statistics
## 
##             Reference
## Prediction   setosa versicolor virginica
##   setosa         35          0         0
##   versicolor      0         35         2
##   virginica       0          0        33
## 
## Overall Statistics
##                                           
##                Accuracy : 0.981           
##                  95% CI : (0.9329, 0.9977)
##     No Information Rate : 0.3333          
##     P-Value [Acc &gt; NIR] : &lt; 2.2e-16       
##                                           
##                   Kappa : 0.9714          
##                                           
##  Mcnemar&#39;s Test P-Value : NA              
## 
## Statistics by Class:
## 
##                      Class: setosa Class: versicolor Class: virginica
## Sensitivity                 1.0000            1.0000           0.9429
## Specificity                 1.0000            0.9714           1.0000
## Pos Pred Value              1.0000            0.9459           1.0000
## Neg Pred Value              1.0000            1.0000           0.9722
## Prevalence                  0.3333            0.3333           0.3333
## Detection Rate              0.3333            0.3333           0.3143
## Detection Prevalence        0.3333            0.3524           0.3143
## Balanced Accuracy           1.0000            0.9857           0.9714</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="svm.html#cb43-1" aria-hidden="true" tabindex="-1"></a>iris_gini_predict<span class="ot">=</span><span class="fu">predict</span>(iris_svm,iris.testing,<span class="at">type=</span><span class="st">&#39;raw&#39;</span>)</span>
<span id="cb43-2"><a href="svm.html#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">confusionMatrix</span>(iris_gini_predict,iris.testing<span class="sc">$</span>Species)</span></code></pre></div>
<pre><code>## Confusion Matrix and Statistics
## 
##             Reference
## Prediction   setosa versicolor virginica
##   setosa         15          0         0
##   versicolor      0         13         2
##   virginica       0          2        13
## 
## Overall Statistics
##                                           
##                Accuracy : 0.9111          
##                  95% CI : (0.7878, 0.9752)
##     No Information Rate : 0.3333          
##     P-Value [Acc &gt; NIR] : 8.467e-16       
##                                           
##                   Kappa : 0.8667          
##                                           
##  Mcnemar&#39;s Test P-Value : NA              
## 
## Statistics by Class:
## 
##                      Class: setosa Class: versicolor Class: virginica
## Sensitivity                 1.0000            0.8667           0.8667
## Specificity                 1.0000            0.9333           0.9333
## Pos Pred Value              1.0000            0.8667           0.8667
## Neg Pred Value              1.0000            0.9333           0.9333
## Prevalence                  0.3333            0.3333           0.3333
## Detection Rate              0.3333            0.2889           0.2889
## Detection Prevalence        0.3333            0.3333           0.3333
## Balanced Accuracy           1.0000            0.9000           0.9000</code></pre>
</div>
<div id="cell-segmentation-example" class="section level2 hasAnchor" number="6.7">
<h2><span class="header-section-number">6.7</span> Cell segmentation example<a href="svm.html#cell-segmentation-example" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Load required libraries</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="svm.html#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caret)</span>
<span id="cb45-2"><a href="svm.html#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pROC)</span>
<span id="cb45-3"><a href="svm.html#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(e1071)</span></code></pre></div>
<p>Load data</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="svm.html#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(segmentationData)</span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="svm.html#cb47-1" aria-hidden="true" tabindex="-1"></a>segClass <span class="ot">&lt;-</span> segmentationData<span class="sc">$</span>Class</span></code></pre></div>
<p>Extract predictors from segmentationData</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="svm.html#cb48-1" aria-hidden="true" tabindex="-1"></a>segData <span class="ot">&lt;-</span> segmentationData[,<span class="dv">4</span><span class="sc">:</span><span class="dv">59</span>]</span></code></pre></div>
<p>Partition data</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="svm.html#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb49-2"><a href="svm.html#cb49-2" aria-hidden="true" tabindex="-1"></a>trainIndex <span class="ot">&lt;-</span> <span class="fu">createDataPartition</span>(<span class="at">y=</span>segClass, <span class="at">times=</span><span class="dv">1</span>, <span class="at">p=</span><span class="fl">0.5</span>, <span class="at">list=</span>F)</span>
<span id="cb49-3"><a href="svm.html#cb49-3" aria-hidden="true" tabindex="-1"></a>segDataTrain <span class="ot">&lt;-</span> segData[trainIndex,]</span>
<span id="cb49-4"><a href="svm.html#cb49-4" aria-hidden="true" tabindex="-1"></a>segDataTest <span class="ot">&lt;-</span> segData[<span class="sc">-</span>trainIndex,]</span>
<span id="cb49-5"><a href="svm.html#cb49-5" aria-hidden="true" tabindex="-1"></a>segClassTrain <span class="ot">&lt;-</span> segClass[trainIndex]</span>
<span id="cb49-6"><a href="svm.html#cb49-6" aria-hidden="true" tabindex="-1"></a>segClassTest <span class="ot">&lt;-</span> segClass[<span class="sc">-</span>trainIndex]</span></code></pre></div>
<p>Set seeds for reproducibility (optional). We will be trying 9 values of the tuning parameter with 5 repeats of 10 fold cross-validation, so we need the following list of seeds.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="svm.html#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb50-2"><a href="svm.html#cb50-2" aria-hidden="true" tabindex="-1"></a>seeds <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="at">mode =</span> <span class="st">&quot;list&quot;</span>, <span class="at">length =</span> <span class="dv">51</span>)</span>
<span id="cb50-3"><a href="svm.html#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>) seeds[[i]] <span class="ot">&lt;-</span> <span class="fu">sample.int</span>(<span class="dv">1000</span>, <span class="dv">9</span>)</span>
<span id="cb50-4"><a href="svm.html#cb50-4" aria-hidden="true" tabindex="-1"></a>seeds[[<span class="dv">51</span>]] <span class="ot">&lt;-</span> <span class="fu">sample.int</span>(<span class="dv">1000</span>,<span class="dv">1</span>)</span></code></pre></div>
<p>We will pass the twoClassSummary function into model training through <strong>trainControl</strong>. Additionally we would like the model to predict class probabilities so that we can calculate the ROC curve, so we use the <strong>classProbs</strong> option.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="svm.html#cb51-1" aria-hidden="true" tabindex="-1"></a>cvCtrl <span class="ot">&lt;-</span> <span class="fu">trainControl</span>(<span class="at">method =</span> <span class="st">&quot;repeatedcv&quot;</span>, </span>
<span id="cb51-2"><a href="svm.html#cb51-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">repeats =</span> <span class="dv">5</span>,</span>
<span id="cb51-3"><a href="svm.html#cb51-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">number =</span> <span class="dv">10</span>,</span>
<span id="cb51-4"><a href="svm.html#cb51-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">summaryFunction =</span> twoClassSummary,</span>
<span id="cb51-5"><a href="svm.html#cb51-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">classProbs =</span> <span class="cn">TRUE</span>,</span>
<span id="cb51-6"><a href="svm.html#cb51-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">seeds=</span>seeds)</span></code></pre></div>
<p>Tune SVM over the cost parameter. The default grid of cost parameters start at 0.25 and double at each iteration. Choosing <code>tuneLength = 9</code> will give us cost parameters of 0.25, 0.5, 1, 2, 4, 8, 16, 32 and 64. The train function will calculate an appropriate value of sigma (the kernel parameter) from the data.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="svm.html#cb52-1" aria-hidden="true" tabindex="-1"></a>svmTune <span class="ot">&lt;-</span> <span class="fu">train</span>(<span class="at">x =</span> segDataTrain,</span>
<span id="cb52-2"><a href="svm.html#cb52-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y =</span> segClassTrain,</span>
<span id="cb52-3"><a href="svm.html#cb52-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">method =</span> <span class="st">&#39;svmRadial&#39;</span>,</span>
<span id="cb52-4"><a href="svm.html#cb52-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">tuneLength =</span> <span class="dv">9</span>,</span>
<span id="cb52-5"><a href="svm.html#cb52-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">preProc =</span> <span class="fu">c</span>(<span class="st">&quot;center&quot;</span>, <span class="st">&quot;scale&quot;</span>),</span>
<span id="cb52-6"><a href="svm.html#cb52-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">metric =</span> <span class="st">&quot;ROC&quot;</span>,</span>
<span id="cb52-7"><a href="svm.html#cb52-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">trControl =</span> cvCtrl)</span>
<span id="cb52-8"><a href="svm.html#cb52-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-9"><a href="svm.html#cb52-9" aria-hidden="true" tabindex="-1"></a>svmTune</span></code></pre></div>
<pre><code>## Support Vector Machines with Radial Basis Function Kernel 
## 
## 1010 samples
##   56 predictor
##    2 classes: &#39;PS&#39;, &#39;WS&#39; 
## 
## Pre-processing: centered (56), scaled (56) 
## Resampling: Cross-Validated (10 fold, repeated 5 times) 
## Summary of sample sizes: 909, 909, 909, 909, 909, 909, ... 
## Resampling results across tuning parameters:
## 
##   C      ROC        Sens       Spec     
##    0.25  0.8767949  0.8636923  0.6916667
##    0.50  0.8820256  0.8769231  0.6955556
##    1.00  0.8820855  0.8812308  0.7072222
##    2.00  0.8805983  0.8787692  0.7105556
##    4.00  0.8741538  0.8756923  0.6905556
##    8.00  0.8600085  0.8695385  0.6694444
##   16.00  0.8472393  0.8732308  0.6477778
##   32.00  0.8369231  0.8769231  0.6144444
##   64.00  0.8246496  0.8661538  0.6133333
## 
## Tuning parameter &#39;sigma&#39; was held constant at a value of 0.01598045
## ROC was used to select the optimal model using the largest value.
## The final values used for the model were sigma = 0.01598045 and C = 1.</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="svm.html#cb54-1" aria-hidden="true" tabindex="-1"></a>svmTune<span class="sc">$</span>finalModel</span></code></pre></div>
<pre><code>## Support Vector Machine object of class &quot;ksvm&quot; 
## 
## SV type: C-svc  (classification) 
##  parameter : cost C = 1 
## 
## Gaussian Radial Basis kernel function. 
##  Hyperparameter : sigma =  0.0159804470403774 
## 
## Number of Support Vectors : 537 
## 
## Objective Function Value : -388.0962 
## Training error : 0.132673 
## Probability model included.</code></pre>
<p>SVM accuracy profile</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="svm.html#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(svmTune, <span class="at">metric =</span> <span class="st">&quot;ROC&quot;</span>, <span class="at">scales =</span> <span class="fu">list</span>(<span class="at">x =</span> <span class="fu">list</span>(<span class="at">log =</span><span class="dv">2</span>)))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmAccuracyProfileCellSegment"></span>
<img src="05-support-vector-machines_files/figure-html/svmAccuracyProfileCellSegment-1.png" alt="SVM accuracy profile." width="80%" />
<p class="caption">
Figure 6.13: SVM accuracy profile.
</p>
</div>
<p>Test set results</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="svm.html#cb57-1" aria-hidden="true" tabindex="-1"></a>svmPred <span class="ot">&lt;-</span> <span class="fu">predict</span>(svmTune, segDataTest)</span>
<span id="cb57-2"><a href="svm.html#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">confusionMatrix</span>(svmPred, segClassTest)</span></code></pre></div>
<pre><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction  PS  WS
##         PS 578 111
##         WS  72 248
##                                           
##                Accuracy : 0.8186          
##                  95% CI : (0.7934, 0.8419)
##     No Information Rate : 0.6442          
##     P-Value [Acc &gt; NIR] : &lt; 2.2e-16       
##                                           
##                   Kappa : 0.5945          
##                                           
##  Mcnemar&#39;s Test P-Value : 0.004969        
##                                           
##             Sensitivity : 0.8892          
##             Specificity : 0.6908          
##          Pos Pred Value : 0.8389          
##          Neg Pred Value : 0.7750          
##              Prevalence : 0.6442          
##          Detection Rate : 0.5728          
##    Detection Prevalence : 0.6829          
##       Balanced Accuracy : 0.7900          
##                                           
##        &#39;Positive&#39; Class : PS              
## </code></pre>
<p>Get predicted class probabilities</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="svm.html#cb59-1" aria-hidden="true" tabindex="-1"></a>svmProbs <span class="ot">&lt;-</span> <span class="fu">predict</span>(svmTune, segDataTest, <span class="at">type=</span><span class="st">&quot;prob&quot;</span>)</span>
<span id="cb59-2"><a href="svm.html#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(svmProbs)</span></code></pre></div>
<pre><code>##          PS         WS
## 1 0.9015892 0.09841085
## 2 0.9127814 0.08721855
## 3 0.9699164 0.03008361
## 4 0.5828048 0.41719516
## 5 0.8283722 0.17162781
## 6 0.7702593 0.22974067</code></pre>
<p>Build a ROC curve</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="svm.html#cb61-1" aria-hidden="true" tabindex="-1"></a>svmROC <span class="ot">&lt;-</span> <span class="fu">roc</span>(segClassTest, svmProbs[,<span class="st">&quot;PS&quot;</span>])</span></code></pre></div>
<pre><code>## Setting levels: control = PS, case = WS</code></pre>
<pre><code>## Setting direction: controls &gt; cases</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="svm.html#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">auc</span>(svmROC)</span></code></pre></div>
<pre><code>## Area under the curve: 0.8912</code></pre>
<p>Plot ROC curve.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="svm.html#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(svmROC, <span class="at">type =</span> <span class="st">&quot;S&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:svmROCcurveCellSegment"></span>
<img src="05-support-vector-machines_files/figure-html/svmROCcurveCellSegment-1.png" alt="SVM ROC curve for cell segmentation data set." width="80%" />
<p class="caption">
Figure 6.14: SVM ROC curve for cell segmentation data set.
</p>
</div>
<p>Calculate area under ROC curve</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="svm.html#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">auc</span>(svmROC)</span></code></pre></div>
<pre><code>## Area under the curve: 0.8912</code></pre>
</div>
<div id="example---regression" class="section level2 hasAnchor" number="6.8">
<h2><span class="header-section-number">6.8</span> Example - regression<a href="svm.html#example---regression" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This example serves to demonstrate the use of SVMs in regression, but perhaps more importantly, it highlights the power and flexibility of the <a href="http://cran.r-project.org/web/packages/caret/index.html">caret</a> package. Earlier we used <em>k</em>-NN for a regression analysis of the <strong>BloodBrain</strong> dataset (see section 04-nearest-neighbours.Rmd). We will repeat the regression analysis, but this time we will fit a radial kernel SVM. Remarkably, a re-run of this analysis using a completely different type of model, requires changes to only two lines of code.</p>
<p>The pre-processing steps and generation of seeds are identical, therefore if the data were still in memory, we could skip this next block of code:</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="svm.html#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(BloodBrain)</span>
<span id="cb69-2"><a href="svm.html#cb69-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-3"><a href="svm.html#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb69-4"><a href="svm.html#cb69-4" aria-hidden="true" tabindex="-1"></a>trainIndex <span class="ot">&lt;-</span> <span class="fu">createDataPartition</span>(<span class="at">y=</span>logBBB, <span class="at">times=</span><span class="dv">1</span>, <span class="at">p=</span><span class="fl">0.8</span>, <span class="at">list=</span>F)</span>
<span id="cb69-5"><a href="svm.html#cb69-5" aria-hidden="true" tabindex="-1"></a>descrTrain <span class="ot">&lt;-</span> bbbDescr[trainIndex,]</span>
<span id="cb69-6"><a href="svm.html#cb69-6" aria-hidden="true" tabindex="-1"></a>concRatioTrain <span class="ot">&lt;-</span> logBBB[trainIndex]</span>
<span id="cb69-7"><a href="svm.html#cb69-7" aria-hidden="true" tabindex="-1"></a>descrTest <span class="ot">&lt;-</span> bbbDescr[<span class="sc">-</span>trainIndex,]</span>
<span id="cb69-8"><a href="svm.html#cb69-8" aria-hidden="true" tabindex="-1"></a>concRatioTest <span class="ot">&lt;-</span> logBBB[<span class="sc">-</span>trainIndex]</span>
<span id="cb69-9"><a href="svm.html#cb69-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-10"><a href="svm.html#cb69-10" aria-hidden="true" tabindex="-1"></a>transformations <span class="ot">&lt;-</span> <span class="fu">preProcess</span>(descrTrain,</span>
<span id="cb69-11"><a href="svm.html#cb69-11" aria-hidden="true" tabindex="-1"></a>                              <span class="at">method=</span><span class="fu">c</span>(<span class="st">&quot;center&quot;</span>, <span class="st">&quot;scale&quot;</span>, <span class="st">&quot;corr&quot;</span>, <span class="st">&quot;nzv&quot;</span>),</span>
<span id="cb69-12"><a href="svm.html#cb69-12" aria-hidden="true" tabindex="-1"></a>                              <span class="at">cutoff=</span><span class="fl">0.75</span>)</span>
<span id="cb69-13"><a href="svm.html#cb69-13" aria-hidden="true" tabindex="-1"></a>descrTrain <span class="ot">&lt;-</span> <span class="fu">predict</span>(transformations, descrTrain)</span>
<span id="cb69-14"><a href="svm.html#cb69-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-15"><a href="svm.html#cb69-15" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb69-16"><a href="svm.html#cb69-16" aria-hidden="true" tabindex="-1"></a>seeds <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="at">mode =</span> <span class="st">&quot;list&quot;</span>, <span class="at">length =</span> <span class="dv">26</span>)</span>
<span id="cb69-17"><a href="svm.html#cb69-17" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>) seeds[[i]] <span class="ot">&lt;-</span> <span class="fu">sample.int</span>(<span class="dv">1000</span>, <span class="dv">50</span>)</span>
<span id="cb69-18"><a href="svm.html#cb69-18" aria-hidden="true" tabindex="-1"></a>seeds[[<span class="dv">26</span>]] <span class="ot">&lt;-</span> <span class="fu">sample.int</span>(<span class="dv">1000</span>,<span class="dv">1</span>)</span></code></pre></div>
<p>In the arguments to the <code>train</code> function we change <code>method</code> from <code>knn</code> to <code>svmRadialE1071</code>. The <code>tunegrid</code> parameter is replaced with <code>tuneLength = 9</code>. Now we are ready to fit an SVM model.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="svm.html#cb70-1" aria-hidden="true" tabindex="-1"></a>svmTune2 <span class="ot">&lt;-</span> <span class="fu">train</span>(descrTrain,</span>
<span id="cb70-2"><a href="svm.html#cb70-2" aria-hidden="true" tabindex="-1"></a>                 concRatioTrain,</span>
<span id="cb70-3"><a href="svm.html#cb70-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">method=</span>svmRadialE1071,</span>
<span id="cb70-4"><a href="svm.html#cb70-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">tuneLength =</span> <span class="dv">9</span>,</span>
<span id="cb70-5"><a href="svm.html#cb70-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">trControl =</span> <span class="fu">trainControl</span>(<span class="at">method=</span><span class="st">&quot;repeatedcv&quot;</span>,</span>
<span id="cb70-6"><a href="svm.html#cb70-6" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">number =</span> <span class="dv">5</span>,</span>
<span id="cb70-7"><a href="svm.html#cb70-7" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">repeats =</span> <span class="dv">5</span>,</span>
<span id="cb70-8"><a href="svm.html#cb70-8" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">seeds=</span>seeds</span>
<span id="cb70-9"><a href="svm.html#cb70-9" aria-hidden="true" tabindex="-1"></a>                                          )</span>
<span id="cb70-10"><a href="svm.html#cb70-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb70-11"><a href="svm.html#cb70-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-12"><a href="svm.html#cb70-12" aria-hidden="true" tabindex="-1"></a>svmTune2</span></code></pre></div>
<pre><code>## Support Vector Machines with Radial Kernel - e1071 
## 
## 168 samples
##  63 predictor
## 
## No pre-processing
## Resampling: Cross-Validated (5 fold, repeated 5 times) 
## Summary of sample sizes: 136, 134, 134, 134, 134, 134, ... 
## Resampling results across tuning parameters:
## 
##   cost   RMSE       Rsquared   MAE      
##    0.25  0.6161239  0.4980242  0.4610789
##    0.50  0.5734394  0.5332846  0.4316062
##    1.00  0.5465607  0.5505891  0.4178574
##    2.00  0.5300490  0.5661867  0.4035017
##    4.00  0.5248272  0.5712901  0.3985361
##    8.00  0.5244977  0.5718310  0.3980940
##   16.00  0.5244977  0.5718310  0.3980940
##   32.00  0.5244977  0.5718310  0.3980940
##   64.00  0.5244977  0.5718310  0.3980940
## 
## RMSE was used to select the optimal model using the smallest value.
## The final value used for the model was cost = 8.</code></pre>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="svm.html#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(svmTune2)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:rmseCorSVM"></span>
<img src="05-support-vector-machines_files/figure-html/rmseCorSVM-1.png" alt="Root Mean Squared Error as a function of cost." width="100%" />
<p class="caption">
Figure 6.15: Root Mean Squared Error as a function of cost.
</p>
</div>
<p>Use model to predict outcomes, after first pre-processing the test set.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="svm.html#cb73-1" aria-hidden="true" tabindex="-1"></a>descrTest <span class="ot">&lt;-</span> <span class="fu">predict</span>(transformations, descrTest)</span>
<span id="cb73-2"><a href="svm.html#cb73-2" aria-hidden="true" tabindex="-1"></a>test_pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(svmTune2, descrTest)</span></code></pre></div>
<p>Prediction performance can be visualized in a scatterplot.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="svm.html#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qplot</span>(concRatioTest, test_pred) <span class="sc">+</span> </span>
<span id="cb74-2"><a href="svm.html#cb74-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;observed&quot;</span>) <span class="sc">+</span></span>
<span id="cb74-3"><a href="svm.html#cb74-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;predicted&quot;</span>) <span class="sc">+</span></span>
<span id="cb74-4"><a href="svm.html#cb74-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:obsPredConcRatiosSVM"></span>
<img src="05-support-vector-machines_files/figure-html/obsPredConcRatiosSVM-1.png" alt="Concordance between observed concentration ratios and those predicted by radial kernel SVM." width="80%" />
<p class="caption">
Figure 6.16: Concordance between observed concentration ratios and those predicted by radial kernel SVM.
</p>
</div>
<p>We can also measure correlation between observed and predicted values.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="svm.html#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(concRatioTest, test_pred)</span></code></pre></div>
<pre><code>## [1] 0.7274762</code></pre>
</div>
<div id="further-reading" class="section level2 hasAnchor" number="6.9">
<h2><span class="header-section-number">6.9</span> Further reading<a href="svm.html#further-reading" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="http://www-bcf.usc.edu/~gareth/ISL/">An Introduction to Statistical Learning</a></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="nearest-neighbours.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="decision-trees.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
